// @uwdata/mosaic-sql/dist/src/functions/join.js
import { FromNode } from '../ast/from.js';
import { JoinNode } from '../ast/join.js';
import { asNode, asTableRef } from '../util/ast.js';
function tableRef(x) {
    return x instanceof FromNode ? x : asTableRef(x);
}
function makeJoin(left, right, variant, options = {}) {
    if (options.on && options.using) {
        throw new Error('Only one join condition (on or using) can be applied.');
    }
    return new JoinNode(tableRef(left), tableRef(right), variant, options.type, options.on, options.using?.map(c => asNode(c)));
}
/**
 * Create a new cross (cartesian product) join.
 * @param left The left table to join.
 * @param right The right table to join.
 */
export function cross_join(left, right) {
    return makeJoin(left, right, 'CROSS');
}
/**
 * Create a new POSITIONAL join.
 * @param left The left table to join.
 * @param right The right table to join.
 */
export function positional_join(left, right) {
    return makeJoin(left, right, 'POSITIONAL');
}
/**
 * Create a new join.
 * @param left The left table to join.
 * @param right The right table to join.
 * @param options The join options.
 */
export function join(left, right, options) {
    return makeJoin(left, right, options?.on || options?.using ? 'REGULAR' : 'NATURAL', options);
}
/**
 * Create a new ASOF join.
 * @param left The left table to join.
 * @param right The right table to join.
 * @param options The join options.
 */
export function asof_join(left, right, options) {
    if (!(options.on || options.using)) {
        throw new Error('ASOF join requires a join condition.');
    }
    return makeJoin(left, right, 'ASOF', options);
}
