// @uwdata/mosaic-sql/dist/src/functions/operators.js
import { BetweenOpNode, NotBetweenOpNode } from '../ast/between-op.js';
import { BinaryOpNode } from '../ast/binary-op.js';
import { InOpNode } from '../ast/in-op.js';
import { AndNode, OrNode } from '../ast/logical-op.js';
import { UnaryOpNode, UnaryPostfixOpNode } from '../ast/unary-op.js';
import { asNode } from '../util/ast.js';
import { nodeList } from '../util/function.js';
function unaryOp(op, expr) {
    return new UnaryOpNode(op, asNode(expr));
}
function unaryPostfixOp(op, expr) {
    return new UnaryPostfixOpNode(op, asNode(expr));
}
function binaryOp(op, left, right) {
    return new BinaryOpNode(op, asNode(left), asNode(right));
}
/**
 * Logical and (AND) operator.
 * @param clauses The input expressions.
 */
export function and(...clauses) {
    return new AndNode(nodeList(clauses));
}
/**
 * Logical or (OR) operator.
 * @param clauses The input expressions.
 */
export function or(...clauses) {
    return new OrNode(nodeList(clauses));
}
/**
 * Logical not (NOT) operator.
 * @param expr The expression to negate.
 */
export function not(expr) {
    return unaryOp('NOT', expr);
}
/**
 * Null check (IS NULL) operator.
 * @param expr The expression to test.
 */
export function isNull(expr) {
    return unaryPostfixOp('IS NULL', expr);
}
/**
 * Non-null check (IS NOT NULL) operator.
 * @param expr The expression to test.
 */
export function isNotNull(expr) {
    return unaryPostfixOp('IS NOT NULL', expr);
}
/**
 * Bitwise not (~) operator.
 * @param expr The input expression.
 */
export function bitNot(expr) {
    return unaryOp('~', expr);
}
/**
 * Bitwise and (&) operator.
 * @param left The left argument.
 * @param right The right argument.
 */
export function bitAnd(left, right) {
    return binaryOp('&', left, right);
}
/**
 * Bitwise or (|) operator.
 * @param left The left argument.
 * @param right The right argument.
 */
export function bitOr(left, right) {
    return binaryOp('|', left, right);
}
/**
 * Bit shift left (<<) operator.
 * @param left The left argument.
 * @param right The right argument.
 */
export function bitLeft(left, right) {
    return binaryOp('<<', left, right);
}
/**
 * Bit shift right (>>) operator.
 * @param left The left argument.
 * @param right The right argument.
 */
export function bitRight(left, right) {
    return binaryOp('>>', left, right);
}
/**
 * Addition (+) operator.
 * @param left The left argument.
 * @param right The right argument.
 */
export function add(left, right) {
    return binaryOp('+', left, right);
}
/**
 * Subtraction (-) operator.
 * @param left The left argument.
 * @param right The right argument.
 */
export function sub(left, right) {
    return binaryOp('-', left, right);
}
/**
 * Multiplication (*) operator.
 * @param left The left argument.
 * @param right The right argument.
 */
export function mul(left, right) {
    return binaryOp('*', left, right);
}
/**
 * Division (/) operator.
 * @param left The left argument.
 * @param right The right argument.
 */
export function div(left, right) {
    return binaryOp('/', left, right);
}
/**
 * Integer division (//) operator.
 * @param left The left argument.
 * @param right The right argument.
 */
export function idiv(left, right) {
    return binaryOp('//', left, right);
}
/**
 * Modulo (%) operator.
 * @param left The left argument.
 * @param right The right argument.
 */
export function mod(left, right) {
    return binaryOp('%', left, right);
}
/**
 * Exponentiation (**) operator.
 * @param left The left argument.
 * @param right The right argument.
 */
export function pow(left, right) {
    return binaryOp('**', left, right);
}
/**
 * Equality comparision (=) operator.
 * @param left The left argument.
 * @param right The right argument.
 */
export function eq(left, right) {
    return binaryOp('=', left, right);
}
/**
 * Non-equality comparision (<>) operator.
 * @param left The left argument.
 * @param right The right argument.
 */
export function neq(left, right) {
    return binaryOp('<>', left, right);
}
/**
 * Less-than comparision (<) operator.
 * @param left The left argument.
 * @param right The right argument.
 */
export function lt(left, right) {
    return binaryOp('<', left, right);
}
/**
 * Greater-than comparision (>) operator.
 * @param left The left argument.
 * @param right The right argument.
 */
export function gt(left, right) {
    return binaryOp('>', left, right);
}
/**
 * Less-than or equal comparision (<=) operator.
 * @param left The left argument.
 * @param right The right argument.
 */
export function lte(left, right) {
    return binaryOp('<=', left, right);
}
/**
 * Greater-than or equal comparision (>=) operator.
 * @param left The left argument.
 * @param right The right argument.
 */
export function gte(left, right) {
    return binaryOp('>=', left, right);
}
/**
 * Null-inclusive non-equality (IS DISTINCT FROM) operator.
 * @param left The left argument.
 * @param right The right argument.
 */
export function isDistinct(left, right) {
    return binaryOp('IS DISTINCT FROM', left, right);
}
/**
 * Null-inclusive equality (IS NOT DISTINCT FROM) operator.
 * @param left The left argument.
 * @param right The right argument.
 */
export function isNotDistinct(left, right) {
    return binaryOp('IS NOT DISTINCT FROM', left, right);
}
/**
 * Range inclusion (BETWEEN) operator.
 * @param expr The expression to test.
 * @param extent The range extent.
 */
export function isBetween(expr, extent) {
    return new BetweenOpNode(asNode(expr), extent?.map(asNode));
}
/**
 * Range exclusion (NOT BETWEEN) operator.
 * @param expr The expression to test.
 * @param extent The range extent.
 */
export function isNotBetween(expr, extent) {
    return new NotBetweenOpNode(asNode(expr), extent?.map(asNode));
}
/**
 * Set inclusion (IN) operator.
 * @param expr The expression to test.
 * @param values The included values.
 */
export function isIn(expr, values) {
    return new InOpNode(asNode(expr), values.map(asNode));
}
