// @cosmograph/ui/modules/legend/color/type/index.js
var v = Object.defineProperty;
var y = (_, r, t) => r in _ ? v(_, r, { enumerable: !0, configurable: !0, writable: !0, value: t }) : _[r] = t;
var s = (_, r, t) => y(_, typeof r != "symbol" ? r + "" : r, t);
import { defaultTypeColorLegendConfig as p, defaultMaxDisplayedItems as m } from "./config.js";
import i from "./style.module.css.js";
import M from "../../../../styles.module.css.js";
import { merge as b } from "../../../../utils.js";
import { DisplayStateManager as g } from "../../../_abstract/display-state-manager/index.js";
class B extends g {
  /**
   * Constructs a new TypeColorLegend instance with the provided container node and optional configuration.
   *
   * @param containerNode - The HTML element that will contain the legend.
   * @param config - Optional configuration object for the legend.
   */
  constructor(t, e) {
    const l = b(p, e ?? {});
    super(t, g.extractConfig(l, p));
    s(this, "_containerNode");
    s(this, "_config");
    s(this, "_scrollContainer");
    s(this, "_legendItems");
    s(this, "_showMoreButton");
    s(this, "_wrapper");
    s(this, "_itemHeight", 0);
    s(this, "_itemMargin", 0);
    s(this, "_visibleItemsCount", 0);
    s(this, "_startIndex", 0);
    s(this, "_endIndex", 0);
    s(this, "_listItems", []);
    s(this, "_selectedItem");
    s(this, "_isExpanded", !1);
    s(this, "_isHidden", !1);
    s(this, "_calculatedWidth", 0);
    s(this, "_maxMeasuredWidth", 0);
    s(this, "_maxMeasuredCharLength", 0);
    s(this, "_label");
    s(this, "_handleShowMore", () => {
      var t, e;
      this._isExpanded = !this._isExpanded, this._isExpanded ? (this._containerNode.style.pointerEvents = "all", this._scrollContainer.style.overflowY = "auto", this._showMoreButton.firstChild.textContent = "show less") : (this._containerNode.style.pointerEvents = "none", this._scrollContainer.style.overflowY = "hidden", this._scrollContainer.scrollTop = 0, this._handleShowMoreCount()), (e = (t = this._config).onShowMoreClick) == null || e.call(t, this._isExpanded), this._calculateOptimalWidth(), this._updateVisibleItems();
    });
    s(this, "_handleShowMoreCount", () => {
      const { maxDisplayedItems: t = m } = this._config, e = this._listItems.length - t;
      this._showMoreButton.firstChild.textContent = `${e} more...`;
    });
    s(this, "_handleScroll", () => {
      requestAnimationFrame(() => {
        this._updateVisibleItems();
      });
    });
    this._config = l, this._containerNode = t, this._containerNode.classList.add(i.typeLegend), this._containerNode.classList.add(M.scrollbar), this._createElements();
  }
  /**
   * Gets the configuration object for the TypeColorLegend instance.
   */
  get config() {
    return this._config;
  }
  /**
   * Gets the HTML element that contains the legend items.
   */
  get legendItems() {
    return this._legendItems;
  }
  /**
   * Gets the expanded state of the TypeColorLegend instance.
   */
  get isExpanded() {
    return this._isExpanded;
  }
  /**
   * Sets the configuration for the TypeColorLegend instance.
   *
   * @param config - Optional configuration object for the legend. If not provided, the existing configuration will be used.
   */
  setConfig(t) {
    const e = this._isExpanded;
    this._config = b(p, t ?? {}), this.setDisplayStateConfig(g.extractConfig(this._config, p)), this._isExpanded = (t == null ? void 0 : t.expanded) ?? e, this._updateLegend();
  }
  /**
   * Sets the selected item in the TypeColorLegend instance.
   *
   * @param item - The TypeColorLegendItem to set as the selected item, or undefined to clear the selection.
   */
  setSelectedItem(t) {
    this._selectedItem = t, this._updateVisibleItems();
  }
  /**
   * Resets the scroll position of the legend container to the top and clears the selected item.
   * This updates the visible items in the legend.
   */
  resetList() {
    this._scrollContainer.scrollTop = 0, this._selectedItem = void 0, this._updateVisibleItems();
  }
  /**
   * Gets the configuration object for the TypeColorLegend instance.
   *
   * @returns The configuration object for the TypeColorLegend instance.
   */
  getConfig() {
    return this._config;
  }
  /**
   * Shows the legend container by removing the 'disabled' class and adding the 'enabled' class to the scroll container.
   */
  show() {
    this._isHidden && (this._isHidden = !1, this._containerNode.classList.remove(i.disabled), this._containerNode.classList.add(i.enabled));
  }
  /**
   * Hides the legend container by removing the 'enabled' class and adding the 'disabled' class to the scroll container.
   */
  hide() {
    this._isHidden || (this._isHidden = !0, this._containerNode.classList.remove(i.enabled), this._containerNode.classList.add(i.disabled));
  }
  /**
   * Destroys the TypeColorLegend instance by removing the scroll event listener and clearing the container node's HTML.
   */
  destroy() {
    this._scrollContainer.removeEventListener("scroll", this._handleScroll), this._containerNode.innerHTML = "", super.destroy();
  }
  /**
   * Sets the text content of the fallback div to 'Data loading...' to indicate that data is currently being loaded.
   */
  setLoadingState() {
    super.setLoadingState();
  }
  _updateLegend() {
    const { maxDisplayedItems: t = m, items: e } = this._config, { noDataMessage: l, loadingMessage: h } = this._displayStateConfig, o = !(e != null && e.length) && l === !1 && h === !1;
    if (this._config.hidden || o ? this.hide() : this.show(), e != null && e.length) {
      this.hideState(), this._wrapper.classList.remove(i.disabled), this._wrapper.classList.add(i.enabled);
      const n = this._config.label;
      this._label.textContent = n ?? "", this._label.title = n ?? "", n ? this._label.classList.remove("hidden") : this._label.classList.add("hidden"), this._listItems = e;
      const a = this._listItems.length;
      this._calculateItemDimensions(), this._calculateOptimalWidth(), this._visibleItemsCount = t > a ? a : t;
      const c = this._itemHeight + this._itemMargin, u = Math.ceil(this._visibleItemsCount * c - this._itemMargin);
      this._scrollContainer.style.height = `${u}px`, this._legendItems.style.height = `${Math.ceil(this._listItems.length * c)}px`, this._scrollContainer.style.overflowY = this._isExpanded ? "auto" : "hidden", this._containerNode.style.pointerEvents = this._isExpanded ? "all" : "none", this._updateVisibleItems(), this._showMoreButton.style.display = this._listItems.length > t ? "block" : "none";
    } else
      this.showState(), this._wrapper.classList.remove(i.enabled), this._wrapper.classList.add(i.disabled);
  }
  _updateVisibleItems() {
    const t = this._itemHeight + this._itemMargin, e = 3, l = this._scrollContainer.scrollTop;
    this._startIndex = Math.max(0, Math.floor(l / t) - e);
    const h = this._isExpanded ? Math.ceil(this._scrollContainer.clientHeight / t + e * 2) : this._visibleItemsCount;
    this._endIndex = Math.min(this._startIndex + h, this._listItems.length);
    const o = document.createDocumentFragment();
    for (let n = this._startIndex; n < this._endIndex; n++) {
      const a = this._listItems[n];
      if (a) {
        const c = this._createLegendItem(a);
        c.style.position = "absolute", c.style.top = `${n * t}px`, o.appendChild(c);
      }
    }
    this._legendItems.innerHTML = "", this._legendItems.appendChild(o), this._updateShowMoreButton();
  }
  _createElements() {
    this._wrapper = document.createElement("div"), this._wrapper.classList.add(i.wrapper), this._label = document.createElement("div"), this._label.classList.add("name"), this._wrapper.appendChild(this._label), this._scrollContainer = document.createElement("div"), this._scrollContainer.classList.add("scroll-container"), this._scrollContainer.addEventListener("scroll", this._handleScroll, { passive: !0 }), this._legendItems = document.createElement("div"), this._legendItems.className = "legend-items", this._scrollContainer.appendChild(this._legendItems), this._showMoreButton = this._createShowMoreButton(), this._wrapper.appendChild(this._scrollContainer), this._wrapper.appendChild(this._showMoreButton), this._containerNode.appendChild(this._wrapper), this._isHidden || this.show();
  }
  _createShowMoreButton() {
    const t = document.createElement("div");
    t.classList.add("show-more"), t.addEventListener("click", this._handleShowMore);
    const e = document.createElement("span");
    return t.appendChild(e), t;
  }
  _updateShowMoreButton() {
    this._isExpanded ? this._showMoreButton.firstChild.textContent = "show less" : this._handleShowMoreCount();
  }
  _calculateItemDimensions() {
    const t = this._createLegendItem({ label: "Sample", color: "#000000" });
    t.style.visibility = "hidden", t.style.position = "absolute", this._legendItems.appendChild(t);
    const e = window.getComputedStyle(t);
    this._itemHeight = t.offsetHeight, this._itemMargin = parseFloat(e.marginTop) + parseFloat(e.marginBottom), this._legendItems.removeChild(t);
  }
  _createLegendItem(t) {
    const e = document.createElement("div");
    e.classList.add("legend-item"), e.title = t.label, this._config.alignRight && e.classList.add("reverted"), this._selectedItem && e.classList.add(this._selectedItem.label === t.label ? i.selected : i.unselected), this._config.enableEvents && (this._config.onClick && (e.style.cursor = "pointer", e.addEventListener("click", (o) => this._config.onClick(o, t))), this._config.onHover && (e.addEventListener("mouseenter", (o) => {
      e.classList.add(i.hovered), this._config.onHover(o, t);
    }), e.addEventListener("mouseleave", (o) => {
      e.classList.remove(i.hovered), this._config.onHover(o, t);
    })));
    const l = document.createElement("div");
    l.className = "bullet", l.style.backgroundColor = t.color;
    const h = document.createElement("div");
    return h.className = "label", h.style.whiteSpace = "nowrap", h.textContent = t.label, e.appendChild(l), e.appendChild(h), e;
  }
  _calculateOptimalWidth() {
    var I;
    if (!this._listItems.length)
      return;
    let t = 0;
    for (const d of this._listItems)
      d.label && d.label.length > t && (t = d.label.length);
    if (this._listItems.length > (this._config.maxDisplayedItems || m)) {
      const d = this._listItems.length - (this._config.maxDisplayedItems || m);
      t = Math.max(t, `${d} more...`.length, 9);
    }
    if (t <= this._maxMeasuredCharLength) {
      this._maxMeasuredWidth && this._applyCalculatedWidth(this._maxMeasuredWidth);
      return;
    }
    this._maxMeasuredCharLength = t;
    const e = getComputedStyle(this._containerNode).getPropertyValue("--cosmograph-color-legend-type-font-size") || "11px", l = getComputedStyle(this._containerNode).getPropertyValue("--cosmograph-color-legend-type-font-family") || "inherit", h = "300", n = document.createElement("canvas").getContext("2d");
    n.font = `${h} ${e} ${l}`;
    let a = 0;
    if (this._listItems.forEach((d) => {
      const x = n.measureText(d.label).width;
      x > a && (a = x);
    }), this._listItems.length > (this._config.maxDisplayedItems || m)) {
      const d = this._listItems.length - (this._config.maxDisplayedItems || m);
      a = Math.max(a, n.measureText(`${d} more...`).width, n.measureText("show less").width);
    }
    const f = 8 + 5 + a + 10, C = ((I = this._containerNode.parentElement) == null ? void 0 : I.offsetWidth) || 0, w = C > 0 ? Math.min(C - 20, f) : f;
    this._calculatedWidth = Math.max(100, w), this._maxMeasuredWidth = this._calculatedWidth, this._applyCalculatedWidth(this._calculatedWidth);
  }
  _applyCalculatedWidth(t) {
    this._containerNode.style.minWidth = `${t}px`, this._wrapper.style.minWidth = `${t}px`;
  }
}
export {
  B as TypeColorLegend
};
