// @cosmograph/ui/utils.js
import { format as D } from "d3-format";
const I = (t) => typeof t == "number", O = (t) => typeof t == "function", $ = (t) => t === void 0, N = (t) => t == null, P = (t) => typeof t == "string", T = (t) => Array.isArray(t), F = (t) => t instanceof Object, m = (t) => t.constructor.name !== "Function" && t.constructor.name !== "Object", y = (t) => F(t) && !T(t) && !O(t) && !m(t), p = (t, e = /* @__PURE__ */ new Map()) => {
  if (typeof t != "object" || t === null)
    return t;
  if (t instanceof Date)
    return new Date(t.getTime());
  if (t instanceof Array) {
    const r = [];
    e.set(t, r);
    for (const n of t)
      r.push(e.has(n) ? e.get(n) : p(n, e));
    return t;
  }
  if (m(t))
    return t;
  if (t instanceof Object) {
    const r = {};
    e.set(t, r);
    const n = t;
    return Object.keys(t).reduce((s, f) => (s[f] = e.has(n[f]) ? e.get(n[f]) : p(n[f], e), s), r), r;
  }
  return t;
}, S = (t, e, r = /* @__PURE__ */ new Map()) => {
  const n = m(t) ? t : p(t);
  return t === e ? t : r.has(e) ? r.get(e) : (r.set(e, n), Object.keys(e).forEach((s) => {
    y(t[s]) && y(e[s]) ? n[s] = S(t[s], e[s], r) : m(e) ? n[s] = e : n[s] = p(e[s]);
  }), n);
}, _ = (t, e, r) => t >= +e && +t <= +r, x = (t) => {
  const e = getComputedStyle(t);
  let r = t.clientWidth, n = t.clientHeight;
  return n -= parseFloat(e.paddingTop) + parseFloat(e.paddingBottom), r -= parseFloat(e.paddingLeft) + parseFloat(e.paddingRight), { height: n, width: r };
}, A = (t, e, r = /* @__PURE__ */ new Set()) => {
  if (Array.isArray(t)) {
    if (!Array.isArray(e) || t.length !== e.length)
      return !1;
    if (r.has(t))
      return !0;
    r.add(t);
    for (let n = 0; n < t.length; n++)
      if (!A(t[n], e[n], r))
        return !1;
    return !0;
  }
  if (t instanceof Date && e instanceof Date)
    return t.getTime() === e.getTime();
  if (typeof t == "object" && t !== null && e !== null) {
    if (typeof e != "object")
      return !1;
    if (t === e)
      return !0;
    if (Object.keys(t).length !== Object.keys(e).length)
      return !1;
    if (r.has(t))
      return !0;
    r.add(t);
    for (const s in t)
      if (!A(t[s], e[s], r))
        return !1;
    return !0;
  }
  return t === e;
}, L = (t, e = "", r, n) => {
  if (!t)
    return "0";
  let s, f;
  r ? (s = t, f = r) : Math.abs(t) >= 7e11 ? (s = t / 1e12, f = "T") : Math.abs(t) >= 2e8 ? (s = t / 1e9, f = "B") : Math.abs(t) >= 1e6 ? (s = t / 1e6, f = "M") : Math.abs(t) >= 1e3 ? (s = t / 1e3, f = "K") : (s = t, f = "");
  const c = Math.abs(s);
  n === void 0 && (Number.isInteger(c) ? n = 0 : c < 0.1 ? n = 3 : c < 1 ? n = 2 : c < 10 ? n = 1 : n = 0);
  let a = c > 0 && c < 0.5 / Math.pow(10, n);
  if (a) {
    let l = n;
    for (; a && l < 8; )
      l++, a = c < 0.5 / Math.pow(10, l);
    if (!a)
      n = l;
    else
      return `~${e ?? ""}0`;
  }
  let o = t < 0 ? "-" : "";
  return e && (o += e), `${o}${D(`.${n}~f`)(c)}${f}`;
};
function B(t, e, r = e, n = []) {
  if (!t.length || !e.length || !r.length)
    return new Array(e.length).fill(0);
  const s = new Array(e.length).fill(0), f = /* @__PURE__ */ new Map();
  for (const [o, i] of r) {
    f.has(o) || f.set(o, /* @__PURE__ */ new Map());
    for (let l = 0; l < e.length; l++) {
      const [u, g] = e[l];
      if (o === u && i === g) {
        f.get(o).set(i, l);
        break;
      }
    }
  }
  const c = n.length > 0 ? n[n.length - 1] : e[e.length - 1][1], a = [...r].sort((o, i) => o[0] - i[0]);
  for (const o of t) {
    let i = 0, l = a.length - 1;
    for (; i <= l; ) {
      const u = Math.floor((i + l) / 2), [g, h] = a[u], M = h === c;
      if (o >= g && (M ? o <= h : o < h)) {
        const d = f.get(g);
        if (d) {
          const w = d.get(h);
          w !== void 0 && s[w]++;
        }
        break;
      } else o < g ? l = u - 1 : i = u + 1;
    }
  }
  return s;
}
function R(t, e, r) {
  if (!t)
    return [];
  const [n, s] = t, f = Math.abs(+s - +n), c = Number.isFinite(f) && e > 1 ? f / (e - 1) : 1, a = r ?? c;
  if (a === 0)
    return [];
  const o = Array.from({ length: e }, (i, l) => +n + l * a);
  return o[e - 1] = +s, o;
}
function q(t, e, r, n = !1, s = !0) {
  const f = Array.from({ length: e.length - 1 }, (o, i) => [e[i], e[i + 1]]), c = r && n ? f.filter(([o, i]) => {
    const [l, u] = r;
    return o < +l ? !1 : i >= +u ? i > +l && o < +u : !0;
  }) : f, a = B(t, f, c, e);
  return f.map(([o, i], l) => ({
    rangeStart: s ? o : new Date(o),
    rangeEnd: s ? i : new Date(i),
    count: a[l] ?? 0
  }));
}
export {
  p as cloneDeep,
  L as defaultFormatter,
  R as generateBandIntervals,
  x as getInnerDimensions,
  m as isAClassInstance,
  T as isArray,
  _ as isBetween,
  A as isEqual,
  O as isFunction,
  N as isNil,
  I as isNumber,
  F as isObject,
  y as isPlainObject,
  P as isString,
  $ as isUndefined,
  q as mapDataToBars,
  B as mapDataToIntervals,
  S as merge
};
