// @uwdata/mosaic-sql/dist/src/ast/join.js
import { JOIN_CLAUSE } from '../constants.js';
import { FromNode } from './from.js';
export class JoinNode extends FromNode {
    /** The left table to join. */
    left;
    /** The right table to join. */
    right;
    /** The join variant (REGULAR, CROSS, NATURAL, POSITIONAL, ASOF). */
    joinVariant;
    /** The join type (INNER, LEFT, RIGHT, FULL, SEMI, ANTI). */
    joinType;
    /**
     * The join condition as a boolean expression.
     * If specified, *using* should be `undefined`.
     */
    condition;
    /**
     * The join condition as shared columns to match on.
     * If specified, *condition* should be `undefined`.
     */
    using;
    /** A table sample to apply to join output. */
    sample;
    /**
     * Instantiate a join node.
     * @param left The left table to join.
     * @param right The right table to join.
     * @param variant The join variant (REGULAR, CROSS, NATURAL, POSITIONAL, ASOF).
     * @param type The join type (INNER, LEFT, RIGHT, FULL, SEMI, ANTI).
     * @param condition The join condition as a boolean expression.
     *  If specified, *using* should be `undefined`.
     * @param using The join condition as shared columns to match on.
     *  If specified, *condition* should be `undefined`.
     * @param sample A table sample to apply to join output.
     */
    constructor(left, right, variant = 'NATURAL', type = 'INNER', condition, using, sample) {
        super(JOIN_CLAUSE);
        this.left = left;
        this.right = right;
        this.joinVariant = variant;
        this.joinType = type;
        this.condition = condition;
        this.using = using;
        this.sample = sample;
    }
}
