// @uwdata/mosaic-sql/dist/src/ast/window-frame.js
import { WINDOW_EXTENT_EXPR, WINDOW_FRAME } from '../constants.js';
import { isParamLike } from '../util/type-check.js';
import { SQLNode } from './node.js';
import { ParamNode } from './param.js';
export const ROWS = 'ROWS';
export const RANGE = 'RANGE';
export const GROUPS = 'GROUPS';
export const PRECEDING = 'PRECEDING';
export const FOLLOWING = 'FOLLOWING';
export const CURRENT_ROW = 'CURRENT ROW';
export const UNBOUNDED = 'UNBOUNDED';
export class WindowFrameNode extends SQLNode {
    /** The frame type, one of ROWS, RANGE, or GROUPS. */
    frameType;
    /** The window frame extent. */
    extent;
    /** The window frame exclusion criteria. */
    exclude;
    /**
     * Instantiate a window frame definition node.
     * @param frameType The frame type, one of ROWS, RANGE, or GROUPS.
     * @param extent The window frame extent.
     * @param exclude The window frame exclusion criteria.
     */
    constructor(frameType, extent, exclude) {
        super(WINDOW_FRAME);
        this.frameType = frameType;
        this.extent = isParamLike(extent) ? new ParamNode(extent) : extent;
        this.exclude = exclude;
    }
}
export class WindowFrameExprNode extends SQLNode {
    /** The window frame extent. */
    scope;
    /**
     * The window frame extent expression. This value should be null
     * in the case of current row or unbounded extent values.
     */
    expr;
    /**
     * Instantiate a window frame definition node.
     * @param scope The frame scope, one of PRECEDING, FOLLOWING, or CURRENT ROW.
     * @param expr The window frame extent expression.
     */
    constructor(scope, expr = null) {
        super(WINDOW_EXTENT_EXPR);
        this.scope = scope;
        this.expr = expr;
    }
}
