// @cosmograph/ui/modules/timeline/index.js
var I = Object.defineProperty;
var T = (g, m, t) => m in g ? I(g, m, { enumerable: !0, configurable: !0, writable: !0, value: t }) : g[m] = t;
var n = (g, m, t) => T(g, typeof m != "symbol" ? m + "" : m, t);
import "d3-transition";
import { ResizeObserver as L } from "@juggle/resize-observer";
import { axisBottom as D } from "d3-axis";
import { brushX as P } from "d3-brush";
import { scaleLinear as x, scaleTime as G, scaleSymlog as k } from "d3-scale";
import { select as p } from "d3-selection";
import { extent as R } from "d3-array";
import { play as M, pause as W } from "./assets.js";
import { defaultTimelineConfig as S } from "./config.js";
import { defaultDateFormat as A } from "./format.js";
import { merge as C, mapDataToBars as y, generateBandIntervals as N, getInnerDimensions as w } from "../../utils.js";
import { DisplayStateManager as b } from "../_abstract/display-state-manager/index.js";
import o from "./style.module.css.js";
class Z extends b {
  constructor(t, i) {
    const e = C(S, i ?? {});
    super(t, b.extractConfig(e, S));
    n(this, "_config");
    n(this, "_animationInterval");
    n(this, "_isAnimationRunning", !1);
    n(this, "_svgParser", new DOMParser());
    n(this, "_svg");
    n(this, "_animationControlDiv");
    n(this, "_playButtonSvg");
    n(this, "_pauseButtonSvg");
    n(this, "_containerNode");
    n(this, "_axisGroup");
    n(this, "_barsGroup");
    n(this, "_brushGroup");
    n(this, "_highlightedBarsGroup");
    n(this, "_resizeObserver");
    n(this, "_height", 0);
    n(this, "_width", 0);
    n(this, "_timelineWidth", 0);
    n(this, "_timelineHeight", 0);
    n(this, "_barWidth", 0);
    n(this, "_maxCount", 0);
    n(this, "_barsData", []);
    n(this, "_highlightedBarsData", []);
    n(this, "_highlightedData");
    n(this, "_timeData");
    n(this, "_extent");
    n(this, "_bandIntervals", []);
    n(this, "_currentSelection");
    n(this, "_currentSelectionInPixels");
    n(this, "_isNumericTimeline", !1);
    n(this, "_firstRender", !0);
    n(this, "_yScale", x());
    n(this, "_timeScale", G());
    n(this, "_numScale", x());
    n(this, "_activeAxisScale", this._timeScale);
    n(this, "_timeAxis", D(this._timeScale));
    n(this, "_numAxis", D(this._numScale));
    n(this, "_brushInstance");
    n(this, "_toggleAnimation", (t) => {
      t.preventDefault(), this._isAnimationRunning ? this.pauseAnimation() : this.playAnimation();
    });
    n(this, "_disableAnimation", () => {
      var t, i;
      this.pauseAnimation(), (t = this._animationControlDiv) == null || t.removeEventListener("click", this._toggleAnimation), (i = this._animationControlDiv) == null || i.remove();
    });
    n(this, "playAnimation", () => {
      var t, i, e, s;
      clearInterval(this._animationInterval), this._currentSelectionInPixels && (this._animationInterval = setInterval(this._animateSelection, this._config.animationSpeed ?? 0), this._isAnimationRunning = !0, (i = (t = this._config).onAnimationPlay) == null || i.call(t, this._isAnimationRunning, this._currentSelection), (e = this._pauseButtonSvg) == null || e.classList.remove(o.hidden), (s = this._playButtonSvg) == null || s.classList.add(o.hidden));
    });
    n(this, "pauseAnimation", () => {
      var t, i, e, s;
      clearInterval(this._animationInterval), this._isAnimationRunning = !1, (i = (t = this._config).onAnimationPause) == null || i.call(t, this._isAnimationRunning, this._currentSelection), (e = this._pauseButtonSvg) == null || e.classList.add(o.hidden), (s = this._playButtonSvg) == null || s.classList.remove(o.hidden);
    });
    n(this, "stopAnimation", () => {
      var t, i;
      this.pauseAnimation(), this.setSelection(void 0), (i = (t = this._config).onBrush) == null || i.call(t, void 0);
    });
    n(this, "_animateSelection", () => {
      var i, e, s, a;
      const t = this._currentSelectionInPixels;
      t && t[0] !== void 0 && t[1] !== void 0 && (this._setSelectionInPixels([t[0] + this._barWidth, t[1] + this._barWidth]), (e = (i = this._config).onAnimationTick) == null || e.call(i, this._currentSelection), (t[1] === ((s = this._currentSelectionInPixels) == null ? void 0 : s[1]) || ((a = this._currentSelectionInPixels) == null ? void 0 : a[1]) === void 0) && this.stopAnimation());
    });
    n(this, "destroy", () => {
      super.destroy(), this._resizeObserver.disconnect(), this._containerNode.innerHTML = "", clearInterval(this._animationInterval);
    });
    this._config = e, this._containerNode = t, this._containerNode.classList.add(o.timeline), this._animationControlDiv = document.createElement("div"), this._animationControlDiv.classList.add(o.animationControl), this._containerNode.appendChild(this._animationControlDiv), this._initializeSVG(), this._initializeGroups(), this._initAnimationControls(), this._resizeObserver = new L(() => {
      this.resize();
    }), this._resizeObserver.observe(this._containerNode);
  }
  _initializeSVG() {
    this._svg = document.createElementNS("http://www.w3.org/2000/svg", "svg"), this._svg.classList.add(o.timelineSvg), this._containerNode.appendChild(this._svg);
  }
  async _initAnimationControls() {
    await (async () => {
      var i, e;
      if (!this._animationControlDiv.firstChild) {
        const s = this._svgParser.parseFromString(M, "image/svg+xml").firstChild, a = this._svgParser.parseFromString(W, "image/svg+xml").firstChild;
        this._pauseButtonSvg = (i = this._animationControlDiv) == null ? void 0 : i.appendChild(a), this._playButtonSvg = (e = this._animationControlDiv) == null ? void 0 : e.appendChild(s);
      }
    })().then(() => {
      var i, e, s, a, r;
      this._isAnimationRunning ? ((s = this._playButtonSvg) == null || s.classList.add(o.playAnimation, o.hidden), (a = this._pauseButtonSvg) == null || a.classList.add(o.pauseAnimation)) : ((i = this._playButtonSvg) == null || i.classList.add(o.playAnimation), (e = this._pauseButtonSvg) == null || e.classList.add(o.pauseAnimation, o.hidden)), this._currentSelection || (r = this._animationControlDiv) == null || r.classList.add(o.disabled), this._animationControlDiv.addEventListener("click", this._toggleAnimation);
    });
  }
  setLoadingState() {
    super.setLoadingState();
  }
  _initializeGroups() {
    this._axisGroup = p(this._svg).append("g").attr("class", o.axis), this._barsGroup = p(this._svg).append("g").attr("class", o.bars), this._highlightedBarsGroup = p(this._svg).append("g").attr("class", o.bars), this._brushGroup = p(this._svg).append("g").attr("class", o.brush);
  }
  get _barPadding() {
    return this._barWidth * (this._config.barPadding ?? 0);
  }
  getCurrentSelection() {
    return this._currentSelection;
  }
  getCurrentSelectionInPixels() {
    return this._currentSelectionInPixels;
  }
  getBarWidth() {
    return this._barWidth - this._barPadding;
  }
  getConfig() {
    return this._config;
  }
  getIsAnimationRunning() {
    return this._isAnimationRunning;
  }
  setConfig(t) {
    var e, s, a, r;
    const i = JSON.parse(JSON.stringify(this._config));
    this._config = C(S, t ?? {}), this.setDisplayStateConfig(b.extractConfig(this._config, S)), i.formatter !== this._config.formatter && this._checkFormatter(), (e = this._config) != null && e.showAnimationControls ? (s = this._animationControlDiv) != null && s.isConnected || this._initAnimationControls() : this._animationControlDiv && this._disableAnimation(), this._config.allowSelection || this._disableBrush(), (((a = this._config) == null ? void 0 : a.dataStep) !== i.dataStep || ((r = this._config) == null ? void 0 : r.barCount) !== i.barCount) && (this._updateTimelineData(this._timeData), this._updateTimelineData(this._highlightedData, !0)), this.resize();
  }
  _checkFormatter() {
    this._config.formatter ? (this._timeAxis.tickFormat((t) => this._config.formatter(t)), this._numAxis.tickFormat((t) => this._config.formatter(t))) : this._extent && (this._timeAxis.tickFormat((t) => A(t, +this._extent[1] - +this._extent[0], this._isNumericTimeline)), this._numAxis.tickFormat((t) => A(t, +this._extent[1] - +this._extent[0], this._isNumericTimeline)));
  }
  _parseData(t) {
    if (!(t != null && t.length))
      return [];
    const i = new Float64Array(t.length);
    let e = 0;
    for (let s = 0; s < t.length; s++) {
      const a = t[s];
      let r;
      a instanceof Date ? r = a.getTime() : r = a, r && !isNaN(r) && (i[e++] = this._normalizeTimeToMilliseconds(r));
    }
    return Array.from(i.subarray(0, e));
  }
  setTimeData(t, i) {
    if (this._highlightedBarsData = [], this._timeData = [], this._currentSelection = void 0, i && i[1] <= i[0])
      return;
    if (this._brushCurrentSelection(), t && t.length > 1) {
      const h = t[0], l = h instanceof Date && !isNaN(h.getTime()), c = typeof h == "number" && new Date(h).getFullYear() > 1970;
      this._isNumericTimeline = !l && !c;
    }
    const e = this._parseData(t), s = i || R(e), r = s[0] < s[1] && e.length > 1;
    p(this._animationControlDiv).style("display", r ? "flex" : "none"), r ? (this.hideState(), this._timeData = e, this._extent = s, this._firstRender = !0) : (this.showState("Empty or invalid timeline data"), this.render()), this._updateTimelineData(this._timeData), this._updateScales(), this.render();
  }
  setHighlightedData(t) {
    this._highlightedBarsData = [], this._timeData && (this._highlightedData = this._parseData(t), this._updateTimelineData(this._highlightedData, !0), this._updateBars(!0));
  }
  _updateTimelineData(t = [], i = !1) {
    const e = i ? "_highlightedBarsData" : "_barsData";
    if (!this._extent) {
      this[e] = [];
      return;
    }
    if (!t.length) {
      this[e] = y([], this._bandIntervals, this._currentSelection, i);
      return;
    }
    i || this._generateBandIntervals();
    const s = y(t, this._bandIntervals, this._currentSelection, i);
    this[e] = s, i || (this._maxCount = Math.max(...s.map((a) => a.count)));
  }
  _generateBandIntervals() {
    if (!this._extent)
      return;
    const { barCount: t = 30, dataStep: i } = this._config;
    this._bandIntervals = N(this._extent, t, i);
  }
  setSelection(t, i = !1) {
    var s, a, r, h, l, c, u;
    const e = this._currentSelection;
    if (t && this._extent) {
      const _ = [
        Math.max(+this._extent[0], Math.min(+t[0], +this._extent[1])),
        Math.min(+this._extent[1], Math.max(+t[1], +this._extent[0]))
      ];
      if (_[0] < _[1]) {
        if (this._config.stickySelection) {
          const d = _.map(this._activeAxisScale);
          this._mapSelection(d);
        } else
          this._currentSelection = _;
        (s = this._animationControlDiv) == null || s.classList.remove(o.disabled);
      } else
        this._currentSelection = void 0, (a = this._animationControlDiv) == null || a.classList.add(o.disabled);
    } else
      this._currentSelection = void 0, (r = this._animationControlDiv) == null || r.classList.add(o.disabled);
    !i && ((e == null ? void 0 : e[0]) !== ((h = this._currentSelection) == null ? void 0 : h[0]) || (e == null ? void 0 : e[1]) !== ((l = this._currentSelection) == null ? void 0 : l[1])) && ((u = (c = this._config).onBrush) == null || u.call(c, this._currentSelection, !0)), this._brushCurrentSelection();
  }
  _setSelectionInPixels(t) {
    var i, e, s, a, r;
    if (t) {
      const h = [
        Math.max(0, Math.min(t[0], this._timelineWidth)),
        Math.min(this._timelineWidth, Math.max(t[1], 0))
      ];
      h[0] < h[1] ? (this._currentSelection = h.map((l) => this._activeAxisScale.invert(l)), (i = this._animationControlDiv) == null || i.classList.remove(o.disabled)) : (this._currentSelection = void 0, (e = this._animationControlDiv) == null || e.classList.add(o.disabled));
    } else
      this._currentSelection = void 0, (s = this._animationControlDiv) == null || s.classList.add(o.disabled);
    (r = (a = this._config).onBrush) == null || r.call(a, this._currentSelection), this._brushCurrentSelection();
  }
  updateDimensions() {
    var s, a, r, h;
    const { height: t, width: i } = w(this._containerNode), { offsetWidth: e } = this._animationControlDiv;
    this._width = i, this._height = t, this._timelineWidth = this._width - (((s = this._config.padding) == null ? void 0 : s.left) ?? 0) - (((a = this._config.padding) == null ? void 0 : a.right) ?? 0) - e, this._timelineHeight = this._height - (((r = this._config.padding) == null ? void 0 : r.top) ?? 0) - (((h = this._config.padding) == null ? void 0 : h.bottom) ?? 0);
  }
  resize() {
    var e, s;
    this.updateDimensions();
    const t = ((e = this._config.padding) == null ? void 0 : e.top) ?? 0, i = ((s = this._config.padding) == null ? void 0 : s.bottom) ?? 0;
    this._timelineHeight > t + i && (this._updateScales(), this._checkLastTickPosition(), this._currentSelection && this.setSelection(this._currentSelection, !0), this.render());
  }
  render() {
    if (this._timelineHeight <= 0 || this._timelineWidth <= 0) {
      this._axisGroup.style("display", "none");
      return;
    }
    this._firstRender && (this.updateDimensions(), this._updateScales(), this._checkLastTickPosition()), this._updateBrush(), this._updateBars(), this._updateBars(!0), this._updateAxis(), this._firstRender && (this._firstRender = !1);
  }
  _updateAxis() {
    if (!this._timeData || !this._timeData.length) {
      this._axisGroup.style("display", "none");
      return;
    }
    const { padding: t, axisTickHeight: i, selectionPadding: e } = this._config;
    this._axisGroup.style("transform", `translate(${((t == null ? void 0 : t.left) ?? 0) + 1}px, ${((t == null ? void 0 : t.top) ?? 0) + (i ?? 0) + 1 + (e ?? 0) / 2}px)`).call(this._isNumericTimeline ? this._numAxis : this._timeAxis).call((s) => s.select(".domain").remove()), this._axisGroup.selectAll(".tick").select("text").attr("class", o.axisTick).attr("y", 0).attr("dy", -(i ?? 0)).attr("dx", "5px"), this._axisGroup.selectAll("line").attr("class", o.axisLine).attr("y2", -(i ?? 0));
  }
  _updateBrush() {
    if (!this._config.allowSelection || this._timelineHeight < 0 || this._timelineWidth < 0) {
      this._brushGroup.selectAll("*").remove();
      return;
    }
    const { selectionPadding: t, padding: i, selectionRadius: e } = this._config, s = this._timelineHeight - (t ?? 0), a = [
      [0, (t ?? 0) / 2],
      [this._timelineWidth, s + (t ?? 0) / 2]
    ];
    this._brushInstance = P().extent(a), this._brushInstance.on("end", ({ selection: r, sourceEvent: h }) => {
      var l, c, u, _, d, v;
      h && (r ? (this._config.stickySelection ? this._mapSelection(r) : this._currentSelection = r.map((f) => this._activeAxisScale.invert(f)), (l = this._animationControlDiv) == null || l.classList.remove(o.disabled), (u = (c = this._config).onBrush) == null || u.call(c, this._currentSelection)) : (this._currentSelection = void 0, (d = (_ = this._config).onBrush) == null || d.call(_, void 0), (v = this._animationControlDiv) == null || v.classList.add(o.disabled)), this._brushCurrentSelection());
    }), this._brushGroup.style("transform", `translate(${(i == null ? void 0 : i.left) ?? 0}px, ${(i == null ? void 0 : i.top) ?? 0}px)`).call(this._brushInstance).call(this._brushInstance.move, this._currentSelectionInPixels || null), this._brushGroup.select("rect.selection").classed(o.selection, !0).attr("rx", e ?? 0).attr("ry", e ?? 0);
  }
  _updateBars(t = !1) {
    const { padding: i, selectionPadding: e } = this._config, s = t ? this._highlightedBarsGroup : this._barsGroup;
    if (!this._timeData || !this._timeData.length) {
      s.selectAll("rect").remove();
      return;
    }
    s.style("transform", `translate(${(i == null ? void 0 : i.left) ?? 0}px, ${((i == null ? void 0 : i.top) ?? 0) - (e ?? 0) / 2}px)`);
    const a = t ? 0.5 : 1, r = this._createBars(s, t);
    !t && this._config.onBarHover && r.on("mouseover", (h) => {
      var l, c;
      return (c = (l = this._config).onBarHover) == null ? void 0 : c.call(l, h);
    }), this._animateBars(r, t, a);
  }
  _createBars(t, i) {
    return !this._activeAxisScale || !this._yScale || !this._barsData.length ? (t.selectAll("rect").remove(), t.selectAll("rect")) : t.selectAll(`.${i ? o.highlightedBar : o.bar}`).data(i ? this._highlightedBarsData : this._barsData).join("rect").attr("class", i ? o.highlightedBar : o.bar).attr("x", (e) => this._activeAxisScale(e.rangeStart) + this._barPadding / 2).attr("width", this.getBarWidth()).attr("rx", this._config.barRadius ?? 0).attr("ry", this._config.barRadius ?? 0).attr("y", -this._timelineHeight);
  }
  _animateBars(t, i, e) {
    t.transition().duration(this._isAnimationRunning ? 0 : 300).attr("height", (s) => i ? s.count === 0 ? 0 : this._yScale(s.count) : this._yScale(s.count)).style("opacity", (s) => this._yScale(s.count) === (this._config.minBarHeight ?? 0) ? 0.4 : e);
  }
  _updateScales() {
    if (!this._extent || !this._barsData.length || this._timelineHeight < 0 || this._timelineWidth < 0) {
      this._axisGroup.style("display", "none");
      return;
    }
    this._checkFormatter(), this._axisGroup.style("display", "inherit");
    const t = this._barsData[this._barsData.length - 1], e = Math.floor(this._timelineWidth / 100), s = this._barsData.map((f) => f.rangeStart);
    let a;
    if (s.length > e) {
      const f = Math.ceil(s.length / e);
      a = s.filter((H, B) => B % f === 0);
    } else
      a = s;
    (this._isNumericTimeline ? this._numAxis : this._timeAxis).tickValues(a);
    const r = Math.max(0, this._maxCount), h = Math.max(0, this._config.minBarHeight ?? 0), l = Math.max(0, this._timelineHeight - (this._config.barTopMargin ?? 0) - (this._config.selectionPadding ?? 0));
    this._yScale = this._config.useSymlogScale ? k() : x(), this._yScale.range([h, l]).domain([0, r]).clamp(!0);
    const c = [this._extent[0], t.rangeEnd], u = [0, this._timelineWidth], _ = this._isNumericTimeline ? this._numScale : this._timeScale;
    _.domain(c), _.range(u), _.clamp(!0), this._activeAxisScale = _;
    const d = this._barsData[0], v = this._activeAxisScale(d.rangeEnd) - this._activeAxisScale(d.rangeStart);
    this._barWidth = v;
  }
  _disableBrush() {
    var t, i;
    (t = this._brushInstance) == null || t.clear(this._brushGroup), this._currentSelectionInPixels = void 0, this._currentSelection = void 0, this.pauseAnimation(), this._brushGroup.selectAll("*").remove(), this._config.showAnimationControls && ((i = this._animationControlDiv) == null || i.classList.add(o.disabled));
  }
  _mapSelection(t) {
    if (!this._barsData.length)
      return;
    const i = this._activeAxisScale.invert(t[0]), e = this._activeAxisScale.invert(t[1]), s = this._getClosestRange(+i), a = this._getClosestRange(+e, !0);
    s && a && s.rangeStart <= a.rangeEnd && t[1] - t[0] > this.getBarWidth() / 2 ? this._currentSelection = [s.rangeStart, a.rangeEnd] : this._currentSelection = void 0;
  }
  _getClosestRange(t, i) {
    return this._barsData.reduce((e, s) => {
      const a = i ? e.rangeEnd : e.rangeStart, r = i ? s.rangeEnd : s.rangeStart;
      return Math.abs(+r - +t) < Math.abs(+a - +t) ? s : e;
    });
  }
  _checkLastTickPosition() {
    var i;
    const t = this._axisGroup.selectAll(".tick:last-of-type").nodes();
    if (t != null && t.length) {
      const e = t[0], s = e == null ? void 0 : e.getBoundingClientRect().right, a = (i = this._svg) == null ? void 0 : i.getBoundingClientRect().right;
      e.style.display = s >= a ? "none" : "inherit";
    }
  }
  _brushCurrentSelection() {
    var t;
    this._currentSelection ? (this._currentSelectionInPixels = this._currentSelection.map(this._activeAxisScale), this._brushInstance && !this._firstRender && this._brushGroup.call(this._brushInstance.move, this._currentSelectionInPixels)) : (this._currentSelectionInPixels = void 0, (t = this._brushInstance) == null || t.clear(this._brushGroup));
  }
  _normalizeTimeToMilliseconds(t) {
    if (t < 1e13)
      return t;
    const e = 10 ** (Math.floor(Math.log10(t)) - 12);
    return Math.round(t / e * 1e3) / 1e3;
  }
}
export {
  Z as Timeline,
  A as defaultDateFormat
};
