// @cosmograph/ui/modules/legend/color/range/index.js
var E = Object.defineProperty;
var b = (c, d, e) => d in c ? E(c, d, { enumerable: !0, configurable: !0, writable: !0, value: e }) : c[d] = e;
var l = (c, d, e) => b(c, typeof d != "symbol" ? d + "" : d, e);
import { defaultRangeColorLegendConfig as m } from "./config.js";
import o from "./style.module.css.js";
import { merge as x } from "../../../../utils.js";
import { DisplayStateManager as f } from "../../../_abstract/display-state-manager/index.js";
class S extends f {
  /**
   * Constructs a new `RangeColorLegend` instance with the provided container node and optional configuration.
   *
   * The `RangeColorLegend` is responsible for creating and managing the visual elements of the range color legend,
   * including the color gradient, minimum and maximum value labels, and optional event handlers.
   *
   * @param containerNode - The HTML element that will contain the range color legend.
   * @param config - An optional configuration object that can be used to customize the behavior and appearance of the legend.
   */
  constructor(e, t) {
    const n = x(m, t ?? {}), i = f.extractConfig(n, m);
    super(e, i);
    l(this, "_config");
    l(this, "_containerNode");
    l(this, "_legend");
    l(this, "_legendElements");
    l(this, "_gradientElement");
    l(this, "_extentElements", null);
    l(this, "_isHidden", !1);
    l(this, "_handleClick", (e) => {
      var t, n;
      this._config.enableEvents && ((n = (t = this._config).onClick) == null || n.call(t, e));
    });
    l(this, "_handleHover", (e) => {
      var t, n;
      this._config.enableEvents && ((n = (t = this._config).onHover) == null || n.call(t, e));
    });
    l(this, "_handleColorElementClick", (e) => {
      var i, s;
      if (!this._config.enableEvents)
        return;
      const t = e.currentTarget.className.includes("min") ? "min" : "max", n = t === "min" ? this._config.onMinClick : this._config.onMaxClick;
      if (n) {
        const r = t === "min" ? 0 : 1, a = (i = this._config.extent) == null ? void 0 : i[t === "min" ? 0 : this._config.extent.length - 1], h = (s = this._config.extentLabels) == null ? void 0 : s[r];
        n(a, h, e);
      }
    });
    l(this, "_handleColorElementHover", (e) => {
      var h, _, g, u;
      if (!this._config.enableEvents || !this._config.extent)
        return;
      const t = e.currentTarget;
      if (!t)
        return;
      const n = t.classList.contains("min") ? "min" : "max", i = e.type === "mouseenter", s = n === "min" ? 0 : this._config.extent.length - 1, r = (h = this._config.extent) == null ? void 0 : h[s], a = (_ = this._config.extentLabels) == null ? void 0 : _[s];
      t.classList.toggle(o.hovered, i), (u = (g = this._config).onHover) == null || u.call(g, {
        target: t,
        type: n,
        color: r,
        value: a
      });
    });
    this._config = n, this._containerNode = e, this._containerNode.classList.add(o.container), this._createElements();
  }
  /**
   * Gets the minimum and maximum extent elements of the range color legend.
   * @returns The minimum and maximum extent elements, or `null` if they have not been created yet.
   */
  get extentElements() {
    return this._extentElements;
  }
  /**
   * Gets the current configuration for the range color legend.
   * @returns The current configuration object for the range color legend.
   */
  get config() {
    return this._config;
  }
  /**
   * Hides the range color legend by removing the 'enabled' class and adding the 'disabled' class to the legend element.
   */
  hide() {
    this._isHidden || (this._isHidden = !0, this._containerNode.classList.remove(o.enabled), this._containerNode.classList.add(o.disabled));
  }
  /**
   * Shows the range color legend by removing the 'disabled' class and adding the 'enabled' class to the legend element.
   */
  show() {
    this._isHidden && (this._isHidden = !1, this._containerNode.classList.remove(o.disabled), this._containerNode.classList.add(o.enabled));
  }
  /**
   * Sets the configuration for the range color legend.
   *
   * @param config - An optional configuration object that can be used to customize the behavior and appearance of the legend.
   */
  setConfig(e) {
    this._config = x(m, e ?? {}), this.setDisplayStateConfig(f.extractConfig(this._config, m)), this._updateLegend();
  }
  /**
   * Destroys the range color legend by removing event listeners and removing the legend and fallback elements from the container node.
   */
  destroy() {
    this._legend && (this._legend.removeEventListener("click", this._handleClick), this._legend.removeEventListener("mouseover", this._handleHover), this._extentElements && ["min", "max"].forEach((e) => {
      const t = this._extentElements[e];
      t.removeEventListener("click", this._handleColorElementClick), t.removeEventListener("mouseenter", this._handleColorElementHover), t.removeEventListener("mouseleave", this._handleColorElementHover);
    }), this._containerNode.removeChild(this._legend), super.destroy());
  }
  /**
   * Sets the text content of the fallback div to 'Data loading...' to indicate that data is currently being loaded.
   */
  setLoadingState() {
    super.setLoadingState();
  }
  _createElements() {
    this._legend = document.createElement("div"), this._legend.classList.add(o.legend), this._legendElements = this._createLegend(), this._gradientElement = this._createGradientElement(), this._legend.appendChild(this._legendElements), this._legend.appendChild(this._gradientElement), this._containerNode.appendChild(this._legend), this._isHidden || this.show();
  }
  _createLegend() {
    const e = document.createElement("div");
    e.classList.add("legend-content");
    const t = document.createElement("div");
    return t.classList.add("name"), this._extentElements = {
      min: this._createLabelExtent("min"),
      max: this._createLabelExtent("max")
    }, e.appendChild(this._extentElements.min), e.appendChild(t), e.appendChild(this._extentElements.max), e;
  }
  _createLabelExtent(e) {
    const t = document.createElement("div");
    t.classList.add("extent"), t.classList.add(e);
    const n = document.createElement("div");
    return n.classList.add("label"), t.appendChild(n), t.addEventListener("click", this._handleColorElementClick), t.addEventListener("mouseenter", this._handleColorElementHover), t.addEventListener("mouseleave", this._handleColorElementHover), t;
  }
  _createGradientElement() {
    const e = document.createElement("div");
    return e.classList.add("gradient"), e;
  }
  _updateLegend() {
    var i, s, r;
    const { noDataMessage: e, loadingMessage: t } = this._displayStateConfig, n = !this._config.extent && e === !1 && t === !1;
    if (this._config.hidden || n ? this.hide() : this.show(), this._config.extent) {
      this.hideState();
      const a = this._legendElements.querySelector(".name");
      a && (a.textContent = this._config.label ?? "");
      const h = ((i = this._config.extentLabels) == null ? void 0 : i[0]) === ((s = this._config.extentLabels) == null ? void 0 : s[1]);
      if (this._extentElements)
        if (h) {
          if (this._extentElements.min.style.display = "none", this._extentElements.max.style.display = "none", a && ((r = this._config.extentLabels) == null ? void 0 : r[0]) !== void 0) {
            const _ = this._formatLabel(this._config.extentLabels[0]), g = a.textContent ? `${_}
${a.textContent}` : _;
            a.textContent = g;
          }
        } else
          this._updateLabelWrapper("min"), this._updateLabelWrapper("max"), this._extentElements.min.style.display = "block", this._extentElements.max.style.display = "block";
      this._updateGradient(), this._legend.classList.remove(o.disabled), this._legend.classList.add(o.enabled);
    } else
      this.showState(), this._legend.classList.remove(o.enabled), this._legend.classList.add(o.disabled);
  }
  _updateGradient() {
    if (!this._config.extent)
      return;
    const e = this._config.extent.map((t) => this._getColor(t)).filter((t, n, i) => n === 0 || t !== i[n - 1]);
    if (e.length === 1) {
      this._gradientElement.style.backgroundImage = "none", this._gradientElement.style.backgroundColor = e[0];
      return;
    }
    if (this._config.discreteColors) {
      const t = 100 / e.length, n = e.map((i, s) => `${i} ${s * t}% ${(s + 1) * t}%`);
      this._gradientElement.style.backgroundImage = `linear-gradient(to right, ${n.join(", ")})`;
    } else {
      const t = this._config.extent.map((n, i) => `${this._getColor(n)} ${i / (this._config.extent.length - 1) * 100}%`).join(", ");
      this._gradientElement.style.backgroundImage = `linear-gradient(to right, ${t})`;
    }
  }
  _updateLabelWrapper(e) {
    var r;
    const t = this._legendElements.querySelector(`.extent:${e === "min" ? "first-child" : "last-child"}`);
    if (!t)
      return;
    const n = e === "min" ? 0 : 1, i = (r = this._config.extentLabels) == null ? void 0 : r[n], s = t.querySelector(".label");
    s && i !== void 0 && (s.textContent = this._formatLabel(i)), this._updateSublabel(t, e);
  }
  _updateSublabel(e, t) {
    let n = e.querySelector(".sublabel");
    this._config.showSublabels ? (n || (n = document.createElement("div"), n.classList.add("sublabel"), e.appendChild(n)), n.textContent = t === "min" ? this._config.minSubLabel ?? "min" : this._config.maxSubLabel ?? "max") : n && n.remove();
  }
  _formatLabel(e) {
    return typeof e == "string" ? e : this._config.labelFormatter ? this._config.labelFormatter(e) : e.toString();
  }
  _getColor(e) {
    return Array.isArray(e) ? `rgba(${e[0]}, ${e[1]}, ${e[2]}, ${e[3]})` : e;
  }
}
export {
  S as RangeColorLegend
};
