// @cosmograph/ui/modules/search/index.js
var u = Object.defineProperty;
var p = (a, e, t) => e in a ? u(a, e, { enumerable: !0, configurable: !0, writable: !0, value: t }) : a[e] = t;
var o = (a, e, t) => p(a, typeof e != "symbol" ? e + "" : e, t);
import { defaultSearchConfig as h } from "./config.js";
import { merge as d } from "../../utils.js";
import n from "./style.module.css.js";
import m from "../../styles.module.css.js";
class v {
  /**
   * Creates an instance of the Search component.
   *
   * @param container The HTML element that will contain the search component.
   * @param config Optional configuration for the search component.
   */
  constructor(e, t) {
    o(this, "_container");
    o(this, "_inputElement");
    o(this, "_suggestionList");
    o(this, "_dropdownContainer");
    o(this, "_footerContainer", null);
    o(this, "_clearButton");
    o(this, "_config");
    o(this, "_debounceTimeout");
    o(this, "_placeholderElement", null);
    o(this, "_handleFocus");
    o(this, "_handleBlur");
    o(this, "_handleClearClick");
    o(this, "_handleMouseDown");
    this._container = e, this._container.classList.add(m.scrollbar), this._config = d(h, t ?? {});
    const i = document.createElement("div");
    i.className = n.searchInputWrapper, this._container.appendChild(i), this._inputElement = document.createElement("input"), this._inputElement.className = this._config.inputClassName ?? n.searchInput, i.appendChild(this._inputElement), this._createPlaceholderElement(), this._handleFocus = () => this._updatePlaceholderVisibility(), this._handleBlur = () => this._updatePlaceholderVisibility(), this._handleClearClick = () => this.clearInput(), this._inputElement.addEventListener("focus", this._handleFocus), this._inputElement.addEventListener("blur", this._handleBlur), this._clearButton = document.createElement("button"), this._clearButton.className = n.clearButton, this._clearButton.innerHTML = "×", this._clearButton.addEventListener("click", this._handleClearClick), i.appendChild(this._clearButton), this._dropdownContainer = document.createElement("div"), this._dropdownContainer.className = n.dropdownContainer, this._config.openListUpwards && this._dropdownContainer.classList.add(n.upwards), i.appendChild(this._dropdownContainer), this._suggestionList = document.createElement("ul"), this._suggestionList.classList.add(this._config.suggestionListClassName ?? n.searchSuggestions), this._dropdownContainer.appendChild(this._suggestionList), this._footerContainer = document.createElement("div"), this._footerContainer.className = n.footerContainer, this._dropdownContainer.appendChild(this._footerContainer), this._handleKeyDown = this._handleKeyDown.bind(this), this._handleInput = this._handleInput.bind(this), this._inputElement.addEventListener("input", this._handleInput), this._inputElement.addEventListener("keydown", this._handleKeyDown), this._config.disabled && this.setDisabled(!0);
  }
  get suggestionList() {
    return this._suggestionList;
  }
  get dropdownContainer() {
    return this._dropdownContainer;
  }
  _handleInput(e) {
    if (this._config.disabled)
      return;
    const t = this._inputElement.value;
    t ? this._clearButton.classList.add(n.visible) : this._clearButton.classList.remove(n.visible), this._updatePlaceholderVisibility(), this._debounceTimeout && window.clearTimeout(this._debounceTimeout), t.length >= (this._config.minSearchLength ?? 0) && this._renderLoading(), this._debounceTimeout = window.setTimeout(async () => {
      const i = await this._performSearch(t);
      this._config.onInput && this._config.onInput(t, i, e), this._renderSuggestions(i);
    }, this._config.debounceTime ?? 300);
  }
  async _handleKeyDown(e) {
    if (!this._config.disabled && e.key === "Enter") {
      const t = this._inputElement.value;
      this._renderLoading();
      const i = await this._performSearch(t);
      this._config.onEnter && this._config.onEnter(t, i), this._renderSuggestions(i);
    }
  }
  _createPlaceholderElement() {
    var e;
    if (this._config.placeholderRenderer) {
      const t = this._config.placeholderRenderer;
      t instanceof HTMLElement ? (this._placeholderElement = document.createElement("div"), this._placeholderElement.className = n.placeholderWrapper, this._handleMouseDown = (i) => {
        if (this._config.disabled) {
          i.preventDefault();
          return;
        }
        let s = i.target;
        for (; s && s !== this._placeholderElement; ) {
          if (s.dataset.ignoreFocus === "true") {
            i.preventDefault();
            return;
          }
          s = s.parentElement;
        }
      }, this._placeholderElement.addEventListener("mousedown", this._handleMouseDown), this._placeholderElement.appendChild(t), (e = this._inputElement.parentNode) == null || e.insertBefore(this._placeholderElement, this._inputElement)) : this._inputElement.placeholder = t;
    }
    this._updatePlaceholderVisibility();
  }
  _updatePlaceholderVisibility() {
    this._placeholderElement && (this._inputElement.value || document.activeElement === this._inputElement ? this._placeholderElement.classList.add(n.hidden) : this._placeholderElement.classList.remove(n.hidden));
  }
  _createSuggestionElement(e) {
    var l, r;
    const t = document.createElement("li"), i = document.createElement("div"), s = ((r = (l = this._config).suggestionRenderer) == null ? void 0 : r.call(l, e, this._inputElement.value)) ?? String(e);
    return typeof s == "string" ? (i.title = s, i.innerHTML = this._config.highlightMatch ? this.getHighlightedMatch(s) : s) : i.appendChild(s), t.appendChild(i), t;
  }
  _renderSuggestions(e) {
    this._suggestionList.innerHTML = "";
    const t = this._config.maxVisibleItems && e.length > this._config.maxVisibleItems;
    if (this._suggestionList.style.overflowY = t ? "auto" : "", e.length > 0) {
      const l = this._createSuggestionElement(e[0]);
      this._suggestionList.appendChild(l);
      const r = l.offsetHeight;
      if (this._suggestionList.removeChild(l), this._config.maxVisibleItems && r > 0) {
        const c = r * this._config.maxVisibleItems, _ = window.innerHeight * 0.6;
        this._suggestionList.style.maxHeight = `${Math.min(c, _)}px`;
      }
    }
    const s = this._inputElement.value.length >= (this._config.minSearchLength ?? 0);
    if (e.forEach((l) => {
      const r = this._createSuggestionElement(l);
      r.addEventListener("click", () => this._handleSuggestionClick(l)), this._suggestionList.appendChild(r);
    }), e.length === 0 && this._config.noResultsText && s) {
      const l = document.createElement("li");
      l.className = n.noResults, l.textContent = this._config.noResultsText, this._suggestionList.appendChild(l);
    }
    this._footerContainer && (this._footerContainer.innerHTML = "", this._config.footerComponent && s && e.length > 0 && this._footerContainer.appendChild(this._config.footerComponent)), e.length > 0 || this._config.noResultsText && s ? this._dropdownContainer.classList.add(n.visible) : this._dropdownContainer.classList.remove(n.visible);
  }
  _renderLoading() {
    if (!(this._inputElement.value.length >= (this._config.minSearchLength ?? 0)))
      return;
    this._suggestionList.innerHTML = "";
    const i = document.createElement("li");
    i.className = n.loading, i.textContent = "Searching…", this._suggestionList.appendChild(i), this._footerContainer && (this._footerContainer.innerHTML = ""), this._dropdownContainer.classList.add(n.visible);
  }
  _handleSuggestionClick(e) {
    var t, i;
    this._config.disabled || (this._config.onSelect && this._config.onSelect(e), this._inputElement.value = ((i = (t = this._config).selectedSuggestionRenderer) == null ? void 0 : i.call(t, e, this._inputElement.value)) ?? String(e), this._updatePlaceholderVisibility(), this._clearButton.classList.add(n.visible), this._suggestionList.innerHTML = "", this._footerContainer && (this._footerContainer.innerHTML = ""), this._dropdownContainer.classList.remove(n.visible));
  }
  async _performSearch(e) {
    let t = [];
    return this._config.searchFn && e.length >= (this._config.minSearchLength ?? 0) && (t = await this._config.searchFn(e, this._config.suggestionsLimit)), t;
  }
  /**
   * Returns a highlighted match of the input text if {@link SearchConfigInterface.highlightMatch `highlightMatch`} is `true`.
   * Automatically applied if {@link SearchConfigInterface.suggestionRenderer `suggestionRenderer`} is `undefined` or returns string.
   *
   * @param text The text to highlight.
   * @returns The text marked with a <mark> tag of highlighted match or the original text if {@link SearchConfigInterface.highlightMatch `highlightMatch`} is `false` or no match is found.
   */
  getHighlightedMatch(e) {
    if (!this._config.highlightMatch || !this._inputElement.value)
      return e;
    const i = this._inputElement.value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&"), s = new RegExp(`(${i})`, "gi");
    return e.replace(s, "<mark>$1</mark>");
  }
  /**
   * Sets the disabled state of the search component.
   *
   * @param disabled Whether the search component should be disabled.
   */
  setDisabled(e) {
    var t;
    e ? (this._inputElement.disabled = !0, this._inputElement.value = "", (t = this._placeholderElement) == null || t.classList.remove(n.hidden), this._clearButton.classList.remove(n.visible), this._container.classList.add(n.disabled), this._dropdownContainer.classList.remove(n.visible)) : (this._inputElement.disabled = !1, this._container.classList.remove(n.disabled)), this._config.disabled = e;
  }
  /**
   * Returns whether the search component is currently disabled.
   *
   * @returns True if the search component is disabled, false otherwise.
   */
  isDisabled() {
    return !!this._config.disabled;
  }
  /**
   * Sets the configuration for the search component.
   *
   * @param config The new configuration to apply.
   */
  setConfig(e) {
    const t = this._config.openListUpwards;
    this._config = d(h, e ?? {}), this.setDisabled(!!this._config.disabled), t !== this._config.openListUpwards && (this._config.openListUpwards ? this._dropdownContainer.classList.add(n.upwards) : this._dropdownContainer.classList.remove(n.upwards));
  }
  /**
   * Destroys the component and cleans up DOM elements and event listeners.
   */
  destroy() {
    var e;
    this._inputElement.removeEventListener("input", this._handleInput), this._inputElement.removeEventListener("keydown", this._handleKeyDown), this._inputElement.removeEventListener("focus", this._handleFocus), this._inputElement.removeEventListener("blur", this._handleBlur), this._clearButton.removeEventListener("click", this._handleClearClick), this._placeholderElement && this._handleMouseDown && ((e = this._placeholderElement) == null || e.removeEventListener("mousedown", this._handleMouseDown)), this._container.innerHTML = "";
  }
  /**
   * Clears the input and suggestion list.
   */
  clearInput() {
    if (this._config.disabled)
      return;
    const e = !!this._inputElement.value;
    this._inputElement.value = "", this._suggestionList.innerHTML = "", this._footerContainer && (this._footerContainer.innerHTML = ""), this._dropdownContainer.classList.remove(n.visible), this._clearButton.classList.remove(n.visible), this._updatePlaceholderVisibility(), e && this._config.onClear && this._config.onClear();
  }
  /**
   * Hides the suggestion dropdown without altering the current input value.
   */
  hideSuggestions() {
    this._suggestionList.innerHTML = "", this._footerContainer && (this._footerContainer.innerHTML = ""), this._dropdownContainer.classList.remove(n.visible);
  }
  focus() {
    this._inputElement.focus();
  }
  blur() {
    this._inputElement.blur();
  }
}
export {
  v as Search
};
