// @cosmograph/ui/modules/histogram/index.js
var b = Object.defineProperty;
var x = (c, n, t) => n in c ? b(c, n, { enumerable: !0, configurable: !0, writable: !0, value: t }) : c[n] = t;
var r = (c, n, t) => x(c, typeof n != "symbol" ? n + "" : n, t);
import { axisBottom as v } from "d3-axis";
import { ResizeObserver as D } from "@juggle/resize-observer";
import { brushX as y } from "d3-brush";
import { scaleSymlog as B, scaleLinear as C } from "d3-scale";
import { select as u } from "d3-selection";
import { extent as G } from "d3-array";
import { defaultHistogramConfig as d, defaultBarCount as H } from "./config.js";
import { merge as p, defaultFormatter as I, mapDataToBars as m, generateBandIntervals as W } from "../../utils.js";
import { DisplayStateManager as f } from "../_abstract/display-state-manager/index.js";
import h from "./style.module.css.js";
class $ extends f {
  constructor(t, i) {
    const s = p(d, i ?? {}), e = f.extractConfig(s, d);
    super(t, e);
    r(this, "_config");
    r(this, "_containerNode");
    r(this, "_resizeObserver");
    r(this, "_svg");
    r(this, "_axisGroup");
    r(this, "_barsGroup");
    r(this, "_highlightedBarsGroup");
    r(this, "_brushGroup");
    r(this, "_firstRender", !0);
    r(this, "_height");
    r(this, "_width");
    r(this, "_histogramWidth");
    r(this, "_histogramHeight");
    r(this, "_barWidth", 0);
    r(this, "_maxCount", 0);
    r(this, "_extent");
    r(this, "_barsData", []);
    r(this, "_highlightedBarsData", []);
    r(this, "_histogramData");
    r(this, "_highlightedData");
    r(this, "_bandIntervals", []);
    r(this, "_currentSelection");
    r(this, "_currentSelectionInPixels");
    r(this, "_yScale", B());
    r(this, "_xScale", C());
    r(this, "_axis", v(this._xScale));
    r(this, "_brushInstance");
    this._config = s, this._containerNode = t, this._containerNode.classList.add(h.histogram), this._width = t.offsetWidth, this._height = t.offsetHeight, this._histogramWidth = this._width - this._config.padding.left - this._config.padding.right, this._histogramHeight = this._height - this._config.padding.top - this._config.padding.bottom, this._initializeSVG(), this._initializeGroups(), this._resizeObserver = new D(() => this.resize()), this._resizeObserver.observe(this._containerNode);
  }
  _initializeSVG() {
    this._svg = document.createElementNS("http://www.w3.org/2000/svg", "svg"), this._svg.classList.add(h.histogramSvg), this._containerNode.appendChild(this._svg);
  }
  setLoadingState() {
    super.setLoadingState();
  }
  _initializeGroups() {
    this._axisGroup = u(this._svg).append("g").attr("class", h.axis), this._barsGroup = u(this._svg).append("g").attr("class", h.bars), this._highlightedBarsGroup = u(this._svg).append("g").attr("class", h.bars), this._brushGroup = u(this._svg).append("g").attr("class", h.brush);
  }
  get _barPadding() {
    return this._barWidth * (this._config.barPadding ?? 0);
  }
  get _formatter() {
    return this._config.formatter ?? I;
  }
  getCurrentSelection() {
    return this._currentSelection;
  }
  getCurrentSelectionInPixels() {
    return this._currentSelectionInPixels ?? [];
  }
  getBarWidth() {
    return this._barWidth - this._barPadding;
  }
  getConfig() {
    return this._config;
  }
  setConfig(t) {
    const i = JSON.parse(JSON.stringify(this._config));
    this._config = p(d, t ?? {}), this.setDisplayStateConfig(f.extractConfig(this._config, d)), this._config.allowSelection || this._disableBrush(), (this._config.dataStep !== i.dataStep || this._config.barCount !== i.barCount) && (this._updateHistogramData(this._histogramData), this._updateHistogramData(this._highlightedData, !0)), this._axis.tickFormat((s) => this._formatter(s)), this.resize();
  }
  setSelection(t, i = !1) {
    var e, a, o, g;
    const s = this._currentSelection;
    if (t && this._extent) {
      const _ = [
        Math.max(this._extent[0], Math.min(t[0], this._extent[1])),
        Math.min(this._extent[1], Math.max(t[1], this._extent[0]))
      ];
      if (_[0] < _[1])
        if (this._config.stickySelection) {
          const l = _.map(this._xScale);
          this._mapSelection(l);
        } else
          this._currentSelection = _;
      else
        this._currentSelection = void 0;
    } else
      this._currentSelection = void 0;
    !i && ((s == null ? void 0 : s[0]) !== ((e = this._currentSelection) == null ? void 0 : e[0]) || (s == null ? void 0 : s[1]) !== ((a = this._currentSelection) == null ? void 0 : a[1])) && ((g = (o = this._config).onBrush) == null || g.call(o, this._currentSelection, !0)), this._brushCurrentSelection();
  }
  resize() {
    const { offsetWidth: t, offsetHeight: i } = this._containerNode;
    this._width = t, this._height = i, this._histogramWidth = this._width - this._config.padding.left - this._config.padding.right, this._histogramHeight = this._height - this._config.padding.top - this._config.padding.bottom, this._histogramHeight > this._config.padding.top + this._config.padding.bottom && (this._updateScales(), this._currentSelection && this.setSelection(this._currentSelection, !0), this._firstRender || this.render());
  }
  render() {
    this._updateBrush(), this._updateBars(), this._updateBars(!0), this._updateAxis(), this._firstRender && (this._firstRender = !1);
  }
  destroy() {
    super.destroy(), this._resizeObserver.disconnect(), this._containerNode.innerHTML = "";
  }
  _parseData(t) {
    if (!(t != null && t.length))
      return [];
    const i = new Float64Array(t.length);
    let s = 0;
    for (let e = 0; e < t.length; e++) {
      const a = t[e];
      typeof a == "number" && !isNaN(a) && (i[s++] = a);
    }
    return Array.from(i.subarray(0, s));
  }
  setHistogramData(t, i) {
    if (i && i[1] <= i[0])
      return;
    this._histogramData = [], this._highlightedBarsData = [], this._currentSelection = void 0, this._brushCurrentSelection();
    const s = this._parseData(t), e = i || G(s), a = e[0] < e[1];
    a && s.length > 1 ? (this.hideState(), this._histogramData = s, this._extent = e, this._firstRender = !0) : this.showState(a ? void 0 : "Empty or invalid histogram data"), this._updateHistogramData(this._histogramData), this._updateScales(), this.render();
  }
  setHighlightedData(t) {
    this._histogramData && (this._highlightedData = this._parseData(t), this._highlightedBarsData = [], this._updateHistogramData(this._highlightedData, !0), this._updateBars(!0));
  }
  _updateAxis() {
    if (!this._histogramData || !this._histogramData.length) {
      this._axisGroup.style("display", "none");
      return;
    }
    this._axisGroup.style("transform", `translate(${this._config.padding.left}px, ${this._config.padding.top - (this._config.selectionPadding ?? 0) / 2 + 1}px)`).call(this._axis).call((t) => t.select(".domain").remove()), this._axisGroup.selectAll(".tick").select("text").attr("class", h.axisTick).attr("y", 0).attr("dy", -2).attr("dx", this._config.labelSideMargin ?? 0), this._axisGroup.selectAll(".tick:last-of-type text").attr("dx", -(this._config.labelSideMargin ?? 0)).style("text-anchor", "end"), this._axisGroup.selectAll("line").attr("y2", this._histogramHeight).attr("y1", 0).attr("opacity", 0.1);
  }
  _updateBrush() {
    if (!this._config.allowSelection || this._histogramHeight < 0 || this._histogramWidth < 0) {
      this._brushGroup.selectAll("*").remove();
      return;
    }
    const t = this._config.selectionPadding ?? 0, i = this._histogramHeight - t, s = [
      [0, t / 2],
      [this._histogramWidth, i + t / 2]
    ];
    this._brushInstance = y().extent(s), this._brushInstance.on("end", ({ selection: e, sourceEvent: a }) => {
      var o, g, _, l;
      a && (e ? (this._config.stickySelection ? this._mapSelection(e) : this._currentSelection = e.map((S) => this._xScale.invert(S)), (g = (o = this._config).onBrush) == null || g.call(o, this._currentSelection)) : (this._currentSelection = void 0, (l = (_ = this._config).onBrush) == null || l.call(_, void 0)), this._brushCurrentSelection());
    }), this._brushGroup.style("transform", `translate(${this._config.padding.left}px, ${this._config.padding.top}px)`).call(this._brushInstance).call(this._brushInstance.move, this._currentSelectionInPixels || null), this._brushGroup.select("rect.selection").classed(h.selection, !0).attr("rx", this._config.selectionRadius ?? 0).attr("ry", this._config.selectionRadius ?? 0);
  }
  _updateBars(t = !1) {
    const i = t ? this._highlightedBarsGroup : this._barsGroup;
    if (!this._histogramData || !this._histogramData.length) {
      i.selectAll("rect").remove();
      return;
    }
    i.style("transform", `translate(${this._config.padding.left}px, ${this._config.padding.top - (this._config.selectionPadding ?? 0) / 2}px)`);
    const s = t ? 0.5 : 1, e = this._createBars(i, t);
    !t && this._config.onBarHover && e.on("mouseover", this._config.onBarHover), this._animateBars(e, t, s);
  }
  _createBars(t, i) {
    return t.selectAll(`.${i ? h.highlightedBar : h.bar}`).data(i ? this._highlightedBarsData : this._barsData).join("rect").attr("class", i ? h.highlightedBar : h.bar).attr("x", (s) => this._xScale(s.rangeStart) + this._barPadding / 2).attr("width", this.getBarWidth()).attr("rx", this._config.barRadius ?? 0).attr("ry", this._config.barRadius ?? 0).attr("y", -this._histogramHeight);
  }
  _animateBars(t, i, s) {
    t.transition().duration(300).attr("height", (e) => i ? e.count === 0 ? 0 : this._yScale(e.count) : this._yScale(e.count)).style("opacity", (e) => this._yScale(e.count) === this._config.minBarHeight ? 0.4 : s);
  }
  _updateScales() {
    if (!this._extent || !this._barsData.length || this._histogramHeight < 0 || this._histogramWidth < 0)
      this._axisGroup.style("display", "none");
    else {
      this._axisGroup.style("display", "inherit");
      const t = this._barsData[this._barsData.length - 1], i = this._config.dataStep ? [this._extent[0], t.rangeEnd] : this._extent;
      this._xScale.domain(i).range([0, this._histogramWidth]).clamp(!0), this._yScale.range([
        this._config.minBarHeight ?? 0,
        this._histogramHeight - (this._config.barTopMargin ?? 0) - (this._config.selectionPadding ?? 0)
      ]).domain([0, this._maxCount]).clamp(!0), this._axis.tickValues(i), this._barWidth = this._xScale(t.rangeEnd) - this._xScale(t.rangeStart);
    }
  }
  _updateHistogramData(t, i = !1) {
    const s = i ? "_highlightedBarsData" : "_barsData";
    if (!this._extent) {
      this[s] = [];
      return;
    }
    if (!(t != null && t.length)) {
      this[s] = m([], this._bandIntervals, this._currentSelection, i);
      return;
    }
    i || this._generateBandIntervals();
    const e = m(t, this._bandIntervals, this._currentSelection, i);
    this[s] = e, i || (this._maxCount = Math.max(...e.map((a) => a.count)));
  }
  _generateBandIntervals() {
    if (!this._extent)
      return;
    const { barCount: t = H, dataStep: i } = this._config;
    this._bandIntervals = W(this._extent, t, i);
  }
  _disableBrush() {
    this._currentSelection = void 0, this._brushCurrentSelection(), this._brushGroup.selectAll("*").remove();
  }
  _mapSelection(t) {
    if (!this._barsData.length)
      return;
    const i = this._xScale.invert(t[0]), s = this._xScale.invert(t[1]), e = this._getClosestRange(i), a = this._getClosestRange(s, !0);
    e && a && e.rangeStart <= a.rangeEnd && t[1] - t[0] > this.getBarWidth() / 2 ? this._currentSelection = [e.rangeStart, a.rangeEnd] : this._currentSelection = void 0;
  }
  _getClosestRange(t, i) {
    return this._barsData.reduce((s, e) => {
      const a = i ? s.rangeEnd : s.rangeStart, o = i ? e.rangeEnd : e.rangeStart;
      return Math.abs(o - t) < Math.abs(a - t) ? e : s;
    });
  }
  _brushCurrentSelection() {
    var t;
    this._currentSelection ? (this._currentSelectionInPixels = this._currentSelection.map(this._xScale), this._brushInstance && !this._firstRender && this._brushGroup.call(this._brushInstance.move, this._currentSelectionInPixels)) : (this._currentSelectionInPixels = void 0, (t = this._brushInstance) == null || t.clear(this._brushGroup));
  }
}
export {
  $ as Histogram
};
