// @cosmograph/ui/modules/button/index.js
var o = Object.defineProperty;
var s = (n, t, i) => t in n ? o(n, t, { enumerable: !0, configurable: !0, writable: !0, value: i }) : n[t] = i;
var e = (n, t, i) => s(n, typeof t != "symbol" ? t + "" : t, i);
import { merge as c } from "../../utils.js";
import { defaultButtonConfig as r } from "./config.js";
import h from "./style.module.css.js";
class f {
  constructor(t, i) {
    e(this, "_container");
    e(this, "config");
    e(this, "_handleClick", () => {
      var t, i;
      (i = (t = this.config).onClick) == null || i.call(t);
    });
    this._container = t, this._container.classList.add(h.button), this.setConfig(i ?? {}), this._container.addEventListener("click", this._handleClick);
  }
  get element() {
    return this._container;
  }
  setConfig(t) {
    this.config = c(r, t), this._container.title = this.config.description ?? "", this._container.innerHTML = this.config.iconSvg ?? "";
  }
  show() {
    this._container.style.display = "block";
  }
  hide() {
    this._container.style.display = "none";
  }
  destroy() {
    this._container.removeEventListener("click", this._handleClick), this._container.innerHTML = "";
  }
}
export {
  f as Button
};
