// @cosmograph/ui/modules/bars/index.js
var w = Object.defineProperty;
var x = (g, u, t) => u in g ? w(g, u, { enumerable: !0, configurable: !0, writable: !0, value: t }) : g[u] = t;
var n = (g, u, t) => x(g, typeof u != "symbol" ? u + "" : u, t);
import { defaultMaxDisplayedItems as B, defaultBarsConfig as p } from "./config.js";
import L from "./style.module.css.js";
import M from "../../styles.module.css.js";
import { merge as y } from "../../utils.js";
import { DisplayStateManager as v } from "../_abstract/display-state-manager/index.js";
class $ extends v {
  /**
   * Constructs a new instance of the Bars class, which is responsible for rendering a bar chart visualization.
   *
   * @param container - The HTML element that will contain the bar chart.
   * @param config - An optional configuration object that can be used to customize the behavior of the bar chart.
   */
  constructor(t, e) {
    const i = y(p, e ?? {}), s = v.extractConfig(i, p);
    super(t, s);
    n(this, "_data", []);
    n(this, "_originalData", []);
    n(this, "_config");
    n(this, "_container");
    n(this, "_searchContainer");
    n(this, "_searchInput");
    n(this, "_clearButton");
    n(this, "_scrollContainer");
    n(this, "_barsContainer");
    n(this, "_highlightedBarsContainer");
    n(this, "_showMoreButton");
    n(this, "_showMoreCount");
    n(this, "_sortingBlock");
    n(this, "_uniqueValues", /* @__PURE__ */ new Set());
    n(this, "_countMap", new Int32Array(0));
    n(this, "_visibleItemsCount", 0);
    n(this, "_hiddenItemsSum", 0);
    n(this, "_itemHeight", 0);
    n(this, "_itemMargin", 0);
    n(this, "_startIndex", 0);
    n(this, "_endIndex", 0);
    n(this, "_maxCount", 0);
    n(this, "_selectedItem");
    n(this, "_isExpanded", !1);
    n(this, "_sortingHandlers");
    n(this, "_highlightedCounts");
    n(this, "_isFilteringActive", !1);
    n(this, "_updateVisibleItems", () => {
      const t = this._itemHeight + this._itemMargin;
      this._startIndex = Math.floor(this._scrollContainer.scrollTop / t), this._endIndex = Math.min(this._startIndex + this._visibleItemsCount + 1, this._data.length), this._renderItems();
    });
    n(this, "_handleShowMoreCount", () => {
      var i, s, o, a, r, h;
      const t = this._data.length - this._maxDisplayedItems, e = ((s = (i = this._config).countFormatter) == null ? void 0 : s.call(i, this._hiddenItemsSum)) ?? this._hiddenItemsSum.toString();
      if (this._isFilteringActive && this._config.showTotalWhenFiltered) {
        const _ = this._data.slice(this._maxDisplayedItems).reduce((C, f) => C + f.count, 0), d = ((a = (o = this._config).countFormatter) == null ? void 0 : a.call(o, _)) ?? _.toString();
        this._showMoreCount.textContent = `${e} / ${d}`;
      } else
        this._showMoreCount.textContent = e;
      this._showMoreButton.firstChild.textContent = `${((h = (r = this._config).countFormatter) == null ? void 0 : h.call(r, t)) ?? t.toString()} more`;
    });
    n(this, "_handleScroll", () => {
      this._updateVisibleItems();
    });
    n(this, "_handleSort", () => {
      this._config.sort === "alphabetical" ? this.sortAlphabetically(this._config.sortAscending) : this.sortByCount(this._config.sortAscending);
    });
    n(this, "_handleSearch", () => {
      const t = this._searchInput.value.toLowerCase();
      this._clearButton.classList[t ? "add" : "remove"]("visible"), this._data = this._originalData.filter((e) => String(e.label).toLowerCase().includes(t)), this._handleSort();
    });
    n(this, "_clearSearch", () => {
      this._searchInput.value = "", this._handleSearch();
    });
    n(this, "_handleShowMore", () => {
      this._isExpanded = !this._isExpanded, this._isExpanded ? (this._scrollContainer.style.overflowY = "auto", this._showMoreButton.firstChild.textContent = "show less", this._showMoreCount.textContent = "") : (this._scrollContainer.style.overflowY = "hidden", this._scrollContainer.scrollTop = 0, this._handleShowMoreCount());
    });
    this._config = i, this._container = t, this._container.classList.add(L.bars), this._container.classList.add(M.scrollbar), this._scrollContainer = document.createElement("div"), this._scrollContainer.classList.add("scroll-container"), this._barsContainer = document.createElement("div"), this._barsContainer.classList.add("bars-container"), this._highlightedBarsContainer = document.createElement("div"), this._highlightedBarsContainer.classList.add("bars-container", "highlighted-bars-container"), this._scrollContainer.appendChild(this._barsContainer), this._scrollContainer.appendChild(this._highlightedBarsContainer), this._showMoreButton = this._createShowMoreButton(), this._sortingBlock = this._createSortingBlock(), this._searchContainer = this._createSearch(), this._container.appendChild(this._searchContainer), this._container.appendChild(this._scrollContainer), this._container.appendChild(this._showMoreButton), this._container.appendChild(this._sortingBlock), this._scrollContainer.addEventListener("scroll", this._handleScroll, { passive: !0 });
  }
  get _maxDisplayedItems() {
    return this._config.maxDisplayedItems ?? B;
  }
  _createSearch() {
    const t = document.createElement("div");
    return t.classList.add("search-container"), this._clearButton = document.createElement("button"), this._clearButton.classList.add("clear-button"), this._clearButton.innerHTML = "×", this._clearButton.addEventListener("click", this._clearSearch), this._searchInput = document.createElement("input"), this._searchInput.classList.add("search-input"), this._searchInput.placeholder = "Search...", this._searchInput.addEventListener("input", this._handleSearch), t.appendChild(this._searchInput), t.appendChild(this._clearButton), t;
  }
  _updateButtonState(t, e) {
    t.classList.add("selected"), t.classList.remove("unselected"), e.classList.add("unselected"), e.classList.remove("selected");
  }
  _updateArrowText(t, e, i) {
    t.textContent = `${e ? "↑" : "↓"}${i}`;
  }
  _createSortingHandlers(t, e, i, s) {
    return {
      alpha: () => {
        this._config.sort === "alphabetical" && (i = !i, this._updateArrowText(t, i, "ABC")), this._config.sort = "alphabetical", this._config.sortAscending = i, this._updateButtonState(t, e), this._handleSort();
      },
      numeric: () => {
        this._config.sort === "count" && (s = !s, this._updateArrowText(e, s, "123")), this._config.sort = "count", this._config.sortAscending = s, this._updateButtonState(e, t), this._handleSort();
      },
      reset: () => {
        var o, a;
        (a = (o = this._config).onClick) == null || a.call(o, void 0), this.setSelectedItem(void 0);
      }
    };
  }
  _createSortingBlock() {
    const t = document.createElement("div"), e = document.createElement("div"), [i, s] = [document.createElement("div"), document.createElement("div")], o = document.createElement("div");
    t.classList.add("sorting-block"), e.classList.add("buttons-container"), o.classList.add("reset-text", "unselected");
    const a = this._config.sort === "alphabetical" ? this._config.sortAscending : !0, r = this._config.sort === "count" ? this._config.sortAscending : !0;
    return this._updateArrowText(i, a ?? !1, "ABC"), this._updateArrowText(s, r ?? !1, "123"), o.textContent = "reset selection", this._config.sort && this._updateButtonState(this._config.sort === "alphabetical" ? i : s, this._config.sort === "alphabetical" ? s : i), this._sortingHandlers = this._createSortingHandlers(i, s, a ?? !1, r ?? !1), i.addEventListener("click", this._sortingHandlers.alpha), s.addEventListener("click", this._sortingHandlers.numeric), o.addEventListener("click", this._sortingHandlers.reset), e.append(i, s), t.append(e, o), t;
  }
  _createShowMoreButton() {
    const t = document.createElement("div");
    t.classList.add("show-more"), t.addEventListener("click", this._handleShowMore);
    const e = document.createElement("span");
    return this._showMoreCount = document.createElement("span"), this._showMoreCount.classList.add("show-more-count"), t.appendChild(e), t.appendChild(this._showMoreCount), t;
  }
  _calculateVisibleItems() {
    const t = this._itemHeight + this._itemMargin;
    this._visibleItemsCount = Math.ceil(this._scrollContainer.clientHeight / t);
  }
  _createBarElement(t, e = !1, i) {
    var _, d, C, f, b, S, I;
    if (e) {
      const l = document.createElement("div");
      l.classList.add("bar-highlighted"), l.dataset.label = t.label, l.style.position = "absolute";
      const m = i / t.count * (t.count / this._maxCount) * 100;
      return l.style.width = `${m}%`, l;
    }
    const s = document.createElement("div");
    if (s.classList.add("row"), s.style.position = "absolute", (_ = this._config) != null && _.onClick && (s.style.cursor = "pointer", s.addEventListener("click", () => {
      var l, m;
      return (m = (l = this._config).onClick) == null ? void 0 : m.call(l, t);
    })), this._selectedItem) {
      const l = this._container.querySelector(".reset-text");
      l == null || l.classList.remove("unselected"), l == null || l.classList.add("selected"), s.classList.add(this._selectedItem.label === t.label ? "selected" : "unselected");
    }
    const o = t.count / this._maxCount * 100, a = document.createElement("div");
    a.classList.add("bar"), a.style.width = `${o}%`;
    const r = document.createElement("div");
    r.classList.add("info");
    const h = document.createElement("div");
    h.classList.add("label"), h.textContent = t.label, h.title = t.label;
    const c = document.createElement("div");
    if (c.classList.add("count"), this._isFilteringActive && i !== void 0) {
      const l = ((C = (d = this._config).countFormatter) == null ? void 0 : C.call(d, i)) ?? i.toString();
      if (this._config.showTotalWhenFiltered) {
        const m = ((b = (f = this._config).countFormatter) == null ? void 0 : b.call(f, t.count)) ?? t.count.toString();
        c.textContent = `${l} / ${m}`;
      } else
        c.textContent = l;
    } else
      c.textContent = ((I = (S = this._config).countFormatter) == null ? void 0 : I.call(S, t.count)) ?? t.count.toString();
    return r.append(h, c), s.append(a, r), s;
  }
  _renderItems() {
    this._renderItemsToContainer(this._barsContainer, !1), this._renderItemsToContainer(this._highlightedBarsContainer, !0);
  }
  _renderItemsToContainer(t, e) {
    var s;
    const i = this._itemHeight + this._itemMargin;
    if (e) {
      if (!this._highlightedCounts) {
        t.replaceChildren();
        return;
      }
      const o = /* @__PURE__ */ new Map();
      for (const r of Array.from(t.children)) {
        const h = r;
        h.classList.contains("bar-highlighted") && h.dataset.label && o.set(h.dataset.label, h);
      }
      const a = /* @__PURE__ */ new Set();
      for (let r = this._startIndex; r < this._endIndex; r++) {
        const h = this._data[r];
        if (!h)
          continue;
        const c = this._highlightedCounts.get(h.label) || 0;
        if (c > 0) {
          a.add(h.label);
          const _ = o.get(h.label);
          if (_) {
            const d = c / h.count * (h.count / this._maxCount) * 100;
            _.style.width = `${d}%`, _.style.top = `${r * i}px`;
          } else {
            const d = this._createBarElement(h, !0, c);
            d.style.top = `${r * i}px`, d.style.height = `${this._itemHeight}px`, t.appendChild(d);
          }
        }
      }
      for (const [r, h] of o)
        a.has(r) || h.remove();
    } else {
      const o = document.createDocumentFragment();
      for (let a = this._startIndex; a < this._endIndex; a++) {
        const r = this._data[a];
        if (!r)
          continue;
        const h = ((s = this._highlightedCounts) == null ? void 0 : s.get(r.label)) || 0, c = this._createBarElement(r, !1, h);
        c.style.top = `${a * i}px`, c.style.height = `${this._itemHeight}px`, o.appendChild(c);
      }
      t.replaceChildren(o);
    }
  }
  _updateContainerHeights(t) {
    const e = Math.min(this._originalData.length, this._maxDisplayedItems) * t;
    this._scrollContainer.style.height = `${e}px`;
    const i = `${this._data.length * t}px`;
    this._barsContainer.style.height = i, this._highlightedBarsContainer.style.height = i;
  }
  _updateShowMoreButton() {
    if (this._originalData.length > this._maxDisplayedItems) {
      const t = this._data.length > this._maxDisplayedItems;
      this._showMoreButton.style.visibility = t ? "visible" : "hidden", this._showMoreButton.style.display = "flex";
    } else
      this._showMoreButton.style.display = "none";
  }
  _updateUIElements(t = p.showSortingBlock, e = p.showSearch) {
    this._originalData.length ? this._data.length ? (this._sortingBlock.style.visibility = t ? "visible" : "hidden", this._searchContainer.style.display = e ? "block" : "none", this.hideState(), this._scrollContainer.style.visibility = "visible") : (this._sortingBlock.style.visibility = t ? "visible" : "hidden", this._searchContainer.style.display = e ? "block" : "none", this._scrollContainer.style.visibility = "visible", this.showState(this._searchInput.value.length ? "No results found" : void 0)) : (this._sortingBlock.style.visibility = "hidden", this._searchContainer.style.display = "none", this._scrollContainer.style.visibility = "hidden", this._showMoreButton.style.display = "none", this.showState());
  }
  _render() {
    this._maxCount = Math.max(...this._data.map((s) => s.count), 1), this._calculateItemDimensions();
    const { showSortingBlock: t, showSearch: e } = this._config, i = this._itemHeight + this._itemMargin;
    this._updateContainerHeights(i), this._updateShowMoreButton(), this._updateUIElements(t, e), this._calculateVisibleItems(), this._updateVisibleItems();
  }
  _calculateItemDimensions() {
    if (!this._data.length)
      return;
    const t = this._createBarElement(this._data[0]);
    this._barsContainer.appendChild(t);
    const e = window.getComputedStyle(t);
    this._itemHeight = t.offsetHeight, this._itemMargin = parseFloat(e.marginTop) + parseFloat(e.marginBottom), this._barsContainer.removeChild(t);
  }
  _calcHiddenSum() {
    if (this._data.length > this._maxDisplayedItems) {
      const t = this._data.slice(this._maxDisplayedItems);
      this._isFilteringActive ? this._hiddenItemsSum = t.reduce((e, i) => {
        var o;
        const s = ((o = this._highlightedCounts) == null ? void 0 : o.get(i.label)) || 0;
        return e + s;
      }, 0) : this._hiddenItemsSum = t.reduce((e, i) => e + i.count, 0);
    } else
      this._hiddenItemsSum = 0;
    this._handleShowMoreCount();
  }
  /**
   * Returns the current configuration of the bars module.
   * @returns {BarsConfigInterface} The current configuration.
   */
  getConfig() {
    return this._config;
  }
  /**
   * Sets the configuration for the bars module.
   * @param {BarsConfigInterface} [config] - The configuration object to set.
   * @returns {void}
   */
  setConfig(t) {
    this._config = y(p, t ?? {}), this.setDisplayStateConfig(v.extractConfig(this._config, p)), this._isExpanded = (t == null ? void 0 : t.expanded) ?? !1, this._calcHiddenSum(), this._render();
  }
  /**
   * Sets the data for the bars module.
   * @param {any[] | undefined} [values] - The array of values to set as the data.
   * @returns {void}
   */
  setData(t = []) {
    var i, s;
    this._clearSearch(), this._uniqueValues = new Set(t), this._countMap = new Int32Array(this._uniqueValues.size);
    const e = new Map([...this._uniqueValues].map((o, a) => [o, a]));
    t.forEach((o) => {
      const a = e.get(o);
      a !== void 0 && this._countMap[a]++;
    }), this._data = [...this._uniqueValues].map((o, a) => ({
      label: o,
      count: this._countMap[a] || 0
    })), this._originalData = [...this._data], !this._originalData.length && this._selectedItem && (this.setSelectedItem(void 0), (s = (i = this._config).onClick) == null || s.call(i, void 0)), this._searchContainer.style.display = this._originalData.length ? "block" : "none", this._handleSort();
  }
  /**
   * Sorts the data in the bars module by the count of each item, in ascending or descending order.
   * @param {boolean} [ascending=false] - If true, sorts the data in ascending order by count. Otherwise, sorts in descending order.
   * @returns {void}
   */
  sortByCount(t = !1) {
    this._data.sort((e, i) => t ? e.count - i.count : i.count - e.count), this._calcHiddenSum(), this._render();
  }
  /**
   * Sorts the data in the bars module alphabetically by the label of each item, in ascending or descending order.
   * @param {boolean} [ascending=true] - If true, sorts the data in ascending alphabetical order by label. Otherwise, sorts in descending alphabetical order.
   * @returns {void}
   */
  sortAlphabetically(t = !0) {
    this._data.sort((e, i) => t ? e.label.toString().localeCompare(i.label) : i.label.toString().localeCompare(e.label)), this._calcHiddenSum(), this._render();
  }
  /**
   * Destroys the bars module by removing event listeners, clearing the container, and resetting internal data structures.
   * This method should be called when the bars module is no longer needed to clean up any resources it has allocated.
   */
  destroy() {
    this._scrollContainer.removeEventListener("scroll", this._handleScroll), this._searchInput.removeEventListener("input", this._handleSearch), this._clearButton.removeEventListener("click", this._clearSearch), this._showMoreButton.removeEventListener("click", this._handleShowMore);
    const t = this._sortingBlock.querySelector("div:first-child"), e = this._sortingBlock.querySelector("div:nth-child(2)"), i = this._sortingBlock.querySelector(".reset-text");
    t == null || t.removeEventListener("click", this._sortingHandlers.alpha), e == null || e.removeEventListener("click", this._sortingHandlers.numeric), i == null || i.removeEventListener("click", this._sortingHandlers.reset), this._container.innerHTML = "", this._data = [], this._originalData = [], this._uniqueValues.clear(), this._countMap = new Int32Array(0), this._hiddenItemsSum = 0, this._highlightedCounts = void 0, this._isFilteringActive = !1, super.destroy();
  }
  resetList() {
    this._scrollContainer.scrollTop = 0, this._searchInput.value = "", this._selectedItem = void 0, this._highlightedCounts = void 0, this._isFilteringActive = !1, this._data = [...this._originalData], this._updateVisibleItems();
  }
  /**
   * Gets the currently selected item in the bars module.
   * @returns {BarData | undefined} - The currently selected bar data item, or undefined if none selected.
   */
  getSelectedItem() {
    return this._selectedItem;
  }
  /**
   * Sets the currently selected item in the bars module.
   * @param {BarData | undefined} item - The bar data item to select, or undefined to clear the selection.
   * @returns {void}
   */
  setSelectedItem(t) {
    if (!t) {
      const e = this._container.querySelector(".reset-text");
      e == null || e.classList.remove("selected"), e == null || e.classList.add("unselected");
    }
    this._selectedItem = t, this._updateVisibleItems();
  }
  /**
   * Sets the highlighted data for crossfiltering visualization.
   * Creates an overlay bar showing the proportion of filtered data for each category.
   * @param data - Array of string labels that should be highlighted, or undefined to clear highlighting
   */
  setHighlightedData(t) {
    if (t == null)
      this._highlightedCounts = void 0, this._isFilteringActive = !1;
    else {
      this._highlightedCounts = /* @__PURE__ */ new Map();
      for (const i of t)
        this._highlightedCounts.set(i, (this._highlightedCounts.get(i) || 0) + 1);
      const e = this._originalData.reduce((i, s) => i + s.count, 0);
      this._isFilteringActive = t.length < e;
    }
    this._calcHiddenSum(), this._renderItems();
  }
  /**
   * Sets the loading state of the bars module by updating the text content of a fallback div.
   * This method is typically called when data is being loaded asynchronously to provide a visual indication to the user.
   */
  setLoadingState() {
    super.setLoadingState();
  }
}
export {
  $ as Bars
};
