// @cosmograph/ui/modules/_abstract/display-state-manager/index.js
var l = Object.defineProperty;
var n = (a, t, e) => t in a ? l(a, t, { enumerable: !0, configurable: !0, writable: !0, value: e }) : a[t] = e;
var s = (a, t, e) => n(a, typeof t != "symbol" ? t + "" : t, e);
import o from "./style.module.css.js";
const i = {
  noDataMessage: "No data",
  loadingMessage: "Data loading...",
  disableStateMessages: !1,
  allowPointerEvents: !1
};
class y {
  constructor(t, e) {
    s(this, "_displayStateDiv");
    s(this, "_displayStateTimeout", null);
    s(this, "_displayStateConfig");
    s(this, "_displayStateContainerNode");
    this._displayStateContainerNode = t, this._displayStateConfig = { ...i, ...e }, getComputedStyle(t).position === "static" && (t.style.position = "relative"), this._displayStateDiv = this._createDisplayStateElement(), this._displayStateContainerNode.appendChild(this._displayStateDiv);
  }
  static extractConfig(t, e) {
    return {
      noDataMessage: t.noDataMessage ?? e.noDataMessage,
      loadingMessage: t.loadingMessage ?? e.loadingMessage,
      disableStateMessages: t.disableStateMessages ?? e.disableStateMessages,
      allowPointerEvents: t.allowPointerEvents ?? e.allowPointerEvents
    };
  }
  /**
   * Creates the DisplayState element that displays loading/no data states
   */
  _createDisplayStateElement() {
    const t = document.createElement("div");
    return t.classList.add(o.displayState), this._displayStateConfig.allowPointerEvents === !1 && (this._displayStateConfig.noDataMessage === !1 ? this._displayStateContainerNode.style.pointerEvents = "none" : this._displayStateContainerNode.style.pointerEvents = ""), !this._displayStateConfig.disableStateMessages && this._displayStateConfig.noDataMessage !== !1 && (t.textContent = typeof this._displayStateConfig.noDataMessage == "string" ? this._displayStateConfig.noDataMessage : i.noDataMessage), t;
  }
  /**
   * Sets the DisplayState element to loading state
   */
  setLoadingState() {
    !this._displayStateConfig.disableStateMessages && this._displayStateConfig.loadingMessage !== !1 ? this._displayStateDiv.textContent = typeof this._displayStateConfig.loadingMessage == "string" ? this._displayStateConfig.loadingMessage : i.loadingMessage : this._displayStateDiv.textContent = "";
  }
  /**
   * Shows the DisplayState element with the specified text
   * @param text - Optional text to display, defaults to noDataMessage
   */
  showState(t) {
    this._resetDisplayStateTimeout(), this._displayStateConfig.disableStateMessages ? this._displayStateDiv.textContent = "" : t ? this._displayStateDiv.textContent = t : this._displayStateConfig.noDataMessage !== !1 ? this._displayStateDiv.textContent = typeof this._displayStateConfig.noDataMessage == "string" ? this._displayStateConfig.noDataMessage : i.noDataMessage : this._displayStateDiv.textContent = "", this._displayStateConfig.allowPointerEvents === !1 && (this._displayStateContainerNode.style.pointerEvents = "none"), this._displayStateDiv.style.display !== "flex" && (this._displayStateTimeout = setTimeout(() => {
      this._displayStateDiv.style.display = "flex";
    }, 150));
  }
  /**
   * Hides the DisplayState element
   */
  hideState() {
    this._resetDisplayStateTimeout(), this._displayStateDiv.style.display = "none", this._displayStateContainerNode.style.pointerEvents = "";
  }
  /**
   * Updates the DisplayState configuration
   */
  setDisplayStateConfig(t) {
    this._displayStateConfig = { ...this._displayStateConfig, ...t };
  }
  /**
   * Clears any pending DisplayState animation timeouts
   */
  _resetDisplayStateTimeout() {
    this._displayStateTimeout && (clearTimeout(this._displayStateTimeout), this._displayStateTimeout = null);
  }
  /**
   * Cleanup method for DisplayState elements
   */
  destroy() {
    var t, e;
    this._resetDisplayStateTimeout(), (e = (t = this._displayStateDiv) == null ? void 0 : t.parentNode) == null || e.removeChild(this._displayStateDiv);
  }
}
export {
  y as DisplayStateManager,
  i as defaultDisplayStateConfig
};
