// @cosmograph/ui/modules/legend/size/index.js
var v = Object.defineProperty;
var b = (r, d, t) => d in r ? v(r, d, { enumerable: !0, configurable: !0, writable: !0, value: t }) : r[d] = t;
var a = (r, d, t) => b(r, typeof d != "symbol" ? d + "" : d, t);
import { defaultSizeLegendConfig as m } from "./config.js";
import { merge as E } from "../../../utils.js";
import l from "./style.module.css.js";
import { DisplayStateManager as _ } from "../../_abstract/display-state-manager/index.js";
class C extends _ {
  /**
   * Constructs a new `SizeLegend` instance with the provided container node and optional configuration.
   *
   * @param containerNode - The HTML element that will contain the size legend.
   * @param config - An optional configuration object that will be used to initialize the size legend.
   */
  constructor(t, e) {
    const i = E(m, e ?? {});
    super(t, _.extractConfig(i, m));
    a(this, "_config");
    a(this, "_containerNode");
    a(this, "_legend");
    a(this, "_sizeElements", null);
    a(this, "_isHidden", !1);
    a(this, "_handleClick", (t) => {
      var e, i;
      this._config.enableEvents && ((i = (e = this._config).onClick) == null || i.call(e, t));
    });
    a(this, "_handleHover", (t) => {
      var e, i;
      this._config.enableEvents && ((i = (e = this._config).onHover) == null || i.call(e, t));
    });
    a(this, "_handleSizeElementHover", (t) => {
      var h, g, f, u;
      if (!this._config.enableEvents || !this._config.extent || !this._sizeElements)
        return;
      const e = t.currentTarget;
      if (!e)
        return;
      const i = e.classList.contains("min") ? "min" : "max", n = t.type === "mouseenter", s = i === "min" ? 0 : this._config.extent.length - 1, o = (h = this._config.extent) == null ? void 0 : h[s], c = (g = this._config.extentLabels) == null ? void 0 : g[s];
      e.classList.toggle(l.hovered, n), (u = (f = this._config).onHover) == null || u.call(f, {
        target: e,
        type: i,
        size: o,
        value: c
      });
    });
    a(this, "_handleSizeElementClick", (t) => {
      var n, s;
      if (!this._config.enableEvents)
        return;
      const e = t.currentTarget.className.includes("min") ? "min" : "max", i = e === "min" ? this._config.onMinClick : this._config.onMaxClick;
      if (i) {
        const o = ((n = this._config.extent) == null ? void 0 : n[e === "min" ? 0 : 1]) ?? void 0, c = ((s = this._config.extentLabels) == null ? void 0 : s[e === "min" ? 0 : 1]) ?? o;
        i(o, c, t);
      }
    });
    this._config = i, this._containerNode = t, this._containerNode.classList.add(l.container), this._createElements();
  }
  /**
   * Hides the size legend by removing the 'enabled' class and adding the 'disabled' class to the legend element.
   */
  hide() {
    this._isHidden || (this._isHidden = !0, this._containerNode.classList.remove(l.enabled), this._containerNode.classList.add(l.disabled));
  }
  /**
   * Shows the size legend by removing the 'disabled' class and adding the 'enabled' class to the legend element.
   */
  show() {
    this._isHidden && (this._isHidden = !1, this._containerNode.classList.remove(l.disabled), this._containerNode.classList.add(l.enabled));
  }
  /**
   * Sets the configuration for the size legend.
   *
   * @param config - An optional configuration object that will be used to update the size legend.
   */
  setConfig(t) {
    this._config = E(m, t ?? {}), this.setDisplayStateConfig(_.extractConfig(this._config, m)), this._updateLegend();
  }
  /**
   * Sets the text content of the fallback div to 'Data loading...' to indicate that data is currently being loaded.
   */
  setLoadingState() {
    super.setLoadingState();
  }
  /**
   * Destroys the size legend by removing event listeners and removing the legend and fallback message elements from the container node.
   */
  destroy() {
    this._legend && (this._legend.removeEventListener("click", this._handleClick), this._legend.removeEventListener("mouseover", this._handleHover), this._sizeElements && ["min", "max"].forEach((t) => {
      const e = this._sizeElements[t];
      e.removeEventListener("click", this._handleSizeElementClick), e.removeEventListener("mouseenter", this._handleSizeElementHover), e.removeEventListener("mouseleave", this._handleSizeElementHover);
    }), this._containerNode.removeChild(this._legend), super.destroy());
  }
  /**
   * Gets the min and max size elements used in the size legend.
   * @returns The min and max size elements, or null if they have not been created yet.
   */
  get sizeElements() {
    return this._sizeElements;
  }
  /**
   * Gets the current configuration for the size legend.
   * @returns The size legend configuration.
   */
  get config() {
    return this._config;
  }
  /**
   * Gets whether the size legend is currently hidden.
   * @returns `true` if the size legend is hidden, `false` otherwise.
   */
  get isHidden() {
    return this._isHidden;
  }
  _createElements() {
    this._legend = this._createLegendElement(), this._containerNode.appendChild(this._legend), this._isHidden || this.show();
  }
  _createLegendElement() {
    const t = document.createElement("div");
    t.classList.add(l.legend);
    const e = document.createElement("div");
    e.className = "legend-content", this._sizeElements = {
      min: this._createSizeElement("min"),
      max: this._createSizeElement("max")
    }, e.appendChild(this._sizeElements.min), e.appendChild(this._sizeElements.max), t.appendChild(e);
    const i = document.createElement("div");
    return i.className = "name", t.appendChild(i), t;
  }
  _createSizeElement(t) {
    const e = document.createElement("div");
    e.classList.add(t);
    const i = document.createElement("div");
    e.appendChild(i);
    const n = document.createElement("div");
    return n.className = "label", e.appendChild(n), e.addEventListener("click", this._handleSizeElementClick), e.addEventListener("mouseenter", this._handleSizeElementHover), e.addEventListener("mouseleave", this._handleSizeElementHover), e;
  }
  _updateLegend() {
    const { extent: t, extentLabels: e } = this._config, { noDataMessage: i, loadingMessage: n } = this._displayStateConfig, s = !t && i === !1 && n === !1;
    if (this._config.hidden || s ? this.hide() : this.show(), t != null && t.every(Number.isFinite) && (e != null && e.every(Number.isFinite))) {
      this.hideState();
      const o = this._legend.querySelector(".name");
      o && (o.textContent = this._config.label ?? ""), this._sizeElements && (e[0] === e[1] ? (this._updateSizeElement("min"), this._sizeElements.min.style.display = "block", this._sizeElements.max.style.display = "none") : (this._updateSizeElement("min"), this._updateSizeElement("max"), this._sizeElements.min.style.display = "block", this._sizeElements.max.style.display = "block")), this._legend.classList.remove(l.disabled), this._legend.classList.add(l.enabled);
    } else
      this.showState(), this._legend.classList.remove(l.enabled), this._legend.classList.add(l.disabled);
  }
  _updateSizeElement(t) {
    var c, h;
    if (!this._sizeElements)
      return;
    const e = this._sizeElements[t];
    e.style.pointerEvents = this._config.enableEvents ? "auto" : "none";
    const i = ((c = this._config.extent) == null ? void 0 : c[t === "min" ? 0 : 1]) ?? 1, n = ((h = this._config.extentLabels) == null ? void 0 : h[t === "min" ? 0 : 1]) ?? i, s = e.querySelector("div");
    this._config.form === "point" ? (s.classList.remove("line"), s.classList.add("circle"), s.style.height = s.style.width = `${i * (this._config.sizeMultiplier ?? 1)}px`) : (s.style.width = "", s.classList.remove("circle"), s.classList.add("line"), s.style.height = `${i * (this._config.sizeMultiplier ?? 1)}px`);
    const o = e.querySelector(".label");
    o.textContent = this._formatLabel(n), this._updateSublabel(e, t);
  }
  _updateSublabel(t, e) {
    let i = t.querySelector(".sublabel");
    this._config.showSublabels ? (i || (i = document.createElement("div"), i.className = "sublabel", t.appendChild(i)), i.textContent = e === "min" ? this._config.minSubLabel ?? "min" : this._config.maxSubLabel ?? "max") : i && i.remove();
  }
  _formatLabel(t) {
    return this._config.labelFormatter ? this._config.labelFormatter(t) : t.toString();
  }
}
export {
  C as SizeLegend
};
